---
title: "16s_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(reshape2)
library(phyloseq)
library(rstatix)
library(ggpubr)
library(pbapply)
library(DESeq2)
library(DT)
library(metagenomeSeq)
library(data.table)
source("C:/Users/ctata/ANCOM/scripts/ancom_v2.1.R")

sgColorPalette = c("#84CF04","#01B5BB","#E50E63","#6D7272","#8F389E",
                     "#DF8236","#036B6B","#F1BA2F","#9F832D","#94E804",
                     "#01D4DB","#FAC131","#B0B8B8","#F08C3A","#FF106E",
                     "#B948CC","#05B5B5","#CFAA3A")
```

```{r}
output_data = "results/"
ps_deseq <- readRDS(paste0(output_data, "Filtered/ps_deseq_friedfilt.rds"))
ps_css <- readRDS(paste0(output_data,"Filtered/ps_css_friedfilt.rds"))
ps_no_norm_filt <- readRDS(paste0(output_data,"Filtered/ps_no_norm_friedfilt.rds"))
```

### Cutoff parameters
```{r cutoff}
# min post DADA2 counts to run analysis
min_postDD=20000
# DESeq significant cutoff
deseq_cut=0.05
# metagenomeSeq significant cutoff
mtgseq_cut=0.05
# PERMANOVA pvalue and R2 cutoff for visualization
permanova_pcut=0.05
permanova_cut=0.1
```

### DESeq
```{r}
###Run DESeq proper (not just the normalization but all of it)
dir.create(paste0(output_data, 'DESeq/'))
runDESeq <- function(ps, dcut, time = FALSE){
  if(time == TRUE){
    #Working with time series
    #according to the DeSeq vignette: design including the time factor, and then test using the likelihood ratio test as described
    #the following section, where the time factor is removed in the reduced formula
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype + Within.study.sampling.date) 
  }else{
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype) 
  }
  
  #diagdds <- estimateSizeFactors(diagdds, type = "poscounts")
  diagdds <- DESeq(diagdds,fitType="parametric", betaPrior = FALSE) 
  res = results(diagdds, contrast = c("phenotype", "N", "A"))
  res$padj[is.na(res$padj)] = 1
  sig <- res[res$padj < dcut,]
  if (dim(sig)[1] == 0) 
  {sigtab<- as.data.frame(1, row.names="nothing")
  colnames(sigtab) <- 'padj'}
  else 
  {
    sigtab <- data.frame(sig)
  }
  return(list(res, sigtab))
}


###Running analysis 
###split thedata based on the real 3 timepoints
P1<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 1"], ps_no_norm_filt)
P2<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 2"], ps_no_norm_filt)
P3<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 3"], ps_no_norm_filt)

#several significants 
deseq_res_P1 <- runDESeq(P1, deseq_cut) 
deseq_res_P2 <- runDESeq(P2, deseq_cut)
deseq_res_P3 <- runDESeq(P3, deseq_cut)

# print significant taxa
datatable(deseq_res_P1[[2]])
datatable(deseq_res_P2[[2]])
datatable(deseq_res_P3[[2]])
#"ASV_1669" present twice timepoint 1 and 3 


View(as.data.frame(ps_no_norm_filt@tax_table[intersect(rownames(deseq_res_P1[[2]]), rownames(deseq_res_P2[[2]])), ]))

# save
saveRDS(deseq_res_P1, file=paste0(output_data, "DESeq/deseq_res_P1_adjp", deseq_cut, ".rds"))
saveRDS(deseq_res_P2, file=paste0(output_data, "DESeq/deseq_res_P2_adjp", deseq_cut, ".rds"))
saveRDS(deseq_res_P3, file=paste0(output_data, "DESeq/deseq_res_P3_adjp", deseq_cut, ".rds"))



#Including timepoint in the formula
deseq_res_time <- runDESeq(ps_no_norm_filt, deseq_cut, time = TRUE)
View(as.data.frame(ps_no_norm_filt@tax_table[rownames(deseq_res_time[[2]]), ]))
saveRDS(deseq_res_time, file=paste0(output_data, "DESeq/Deseq_time_interaction_adjp", deseq_cut, ".rds"))
```

### MetagenomeSeq
```{r}
dir.create(paste0(output_data, 'metagenomseq/'))
###Run ZIG model fitting and prediction
run_metagenom_seq<-function(ps,maxit, mcut, time = F){
  p_metag<-phyloseq_to_metagenomeSeq(ps)
  #filtering at least 4 samples 
  p_metag= cumNorm(p_metag, p=0.75)
  normFactor =normFactors(p_metag)
  normFactor =log2(normFactor/median(normFactor) + 1)
  if(time == F){
    mod = model.matrix(~phenotype + normFactor, data = pData(p_metag))
  }
  if(time == T){
    mod = model.matrix(~phenotype + Within.study.sampling.date +normFactor, data = pData(p_metag))
  }
  
  settings =zigControl(maxit =maxit, verbose =FALSE)
  fit =fitZig(obj = p_metag, mod = mod, useCSSoffset = FALSE, control = settings)
  #Note: changed fit$taxa to fit@taxa in light of error (probably from newer metagenomeseq ver.)
  res_fit_nonfiltered <- MRtable(fit, number = length(fit$taxa))
  res_fit<-res_fit_nonfiltered[res_fit_nonfiltered$adjPvalues<mcut,]
  #finally remove the ones that are not with enough samples
  Min_effec_samp<-calculateEffectiveSamples(fit)
  Min_effec_samp<-Min_effec_samp[ names(Min_effec_samp)  %in% rownames(res_fit)] #####there is a bug here 
  #manually removing the ones with "NA"
  res_fit<-res_fit[grep("NA",rownames(res_fit), inv=T),]
  res_fit$Min_sample<-Min_effec_samp
  res_fit<-res_fit[res_fit$`+samples in group 0` >= Min_effec_samp & res_fit$`+samples in group 1` >= Min_effec_samp,]
  return(list(res_fit_nonfiltered, res_fit))
}
P1<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 1"], ps_no_norm_filt)
P2<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 2"], ps_no_norm_filt)
P3<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 3"], ps_no_norm_filt)

zig_res_P1 <- run_metagenom_seq(P1,30, mtgseq_cut) # 30The maximum number of iterations for the expectation-maximization algorithm
zig_res_P2 <- run_metagenom_seq(P2,30, mtgseq_cut)
zig_res_P3 <- run_metagenom_seq(P3,30, mtgseq_cut)

zig_res_time<- run_metagenom_seq(ps_no_norm_filt,30, mtgseq_cut, time = T)

# print significant taxa
datatable(zig_res_P1[[2]])
datatable(zig_res_P2[[2]])
datatable(zig_res_P3[[2]])
datatable(zig_res_time[[2]])

zig_res_P1_filtered <- data.frame(cbind(zig_res_P1[[2]], tax_table(P1)[rownames(zig_res_P1[[2]]),]))
zig_res_P1_filtered$enriched <- ifelse(zig_res_P1_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_P2_filtered <- data.frame(cbind(zig_res_P2[[2]], tax_table(P2)[rownames(zig_res_P2[[2]]), ]))
zig_res_P2_filtered$enriched <- ifelse(zig_res_P2_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_P3_filtered <- data.frame(cbind(zig_res_P3[[2]], tax_table(P3)[rownames(zig_res_P3[[2]]), ]))
zig_res_P3_filtered$enriched <- ifelse(zig_res_P3_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_time_filtered <- data.frame(cbind(zig_res_time[[2]], tax_table(ps_no_norm_filt)[rownames(zig_res_time[[2]]), ]))
zig_res_time_filtered$enriched <- ifelse(zig_res_time_filtered$phenotypeN < 0, "Aut", "Control")


saveRDS(zig_res_P1, file=paste0(output_data, "metagenomseq/zig_res_P1_adjp", mtgseq_cut, ".rds"))
saveRDS(zig_res_P2, file=paste0(output_data, "metagenomseq/zig_res_P2_adjp", mtgseq_cut, ".rds"))
saveRDS(zig_res_P3, file=paste0(output_data, "metagenomseq/zig_res_P3_adjp", mtgseq_cut, ".rds"))
saveRDS(zig_res_time, file=paste0(output_data, "metagenomseq/zig_res_time_adjp", mtgseq_cut, ".rds"))

#do we have any in ESV in common?
#Reduce(intersect, list(rownames(zig_res_P1_filtered),rownames(zig_res_P2_filtered),rownames(zig_res_P3_filtered)))
#rownames(zig_res_P1[[2]])[which(rownames(zig_res_P1[[2]]) %in% c(rownames(zig_res_P2[[2]]), rownames(zig_res_P3[[2]])))]
#rownames(zig_res_P2[[2]])[which(rownames(zig_res_P2[[2]]) %in% rownames(zig_res_P3[[2]]))]

metag_res<-cbind(rownames(zig_res_time[[2]]), as(tax_table(ps_no_norm_filt)[rownames(zig_res_time[[2]]), ], "matrix"))
metag_res_esv <-metag_res
rownames(metag_res) <- NULL

```

### Ancom
```{r}

runAncom <- function(ps){
  metada_ps<-sample_data(ps)
  metada_ps<-as.data.frame(metada_ps)
  metada_ps$HostName <- as.character(metada_ps$Host.Name)
  metada_ps$phenotype <- as.character(metada_ps$phenotype)
  
  metada_ps<-tibble(metada_ps$HostName, metada_ps$phenotype, rows = rownames(metada_ps))
  colnames(metada_ps) <- c("HostName", "phenotype", "Biospecimen.Barcode")
  res_table_filt<-otu_table(ps)
  res_table_filt<-as.data.frame(res_table_filt)
  
  prepro<-feature_table_pre_process(feature_table = res_table_filt, meta_data = metada_ps, sample_var = "Biospecimen.Barcode", group_var = "phenotype", out_cut = 0.05, zero_cut = 0.90, lib_cut = 1000, neg_lb = FALSE)
  
  feature_table = prepro$feature_table # Preprocessed feature table
  meta_data = prepro$meta_data # Preprocessed metadata
  struc_zero = prepro$structure_zeros # Structural zero info
  
  main_var = "phenotype"; p_adj_method = "BH"; alpha = 0.05
  adj_formula = NULL; rand_formula = NULL
  ancom_res <- ANCOM(feature_table, meta_data, struc_zero, main_var, p_adj_method, 
                     alpha, adj_formula, rand_formula)#, control)
  return(ancom_res)
}

ancom.all.filt.0.05 <- runAncom(ps_no_norm_filt)
dir.create(path = "ANCOM")
saveRDS(ancom.all.filt.0.05, file=paste0(output_data, "ANCOM/ancom_res", mtgseq_cut, ".rds"))
#no significant taxa from ANCOM



#Analyze each timepoint independently
#Timepoint 3
ps3<-subset_samples(ps_no_norm_filt, Within.study.sampling.date == "Timepoint 3" )
ancom.all.filt.0.05.3 <- runAncom(ps3)
saveRDS(ancom.all.filt.0.05.3, file=paste0(output_data, "ANCOM/ancom_res3_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.3$out$detected_0.6)
# No sig


#Timepoint 2
ps2<-subset_samples(ps_no_norm_filt, Within.study.sampling.date == "Timepoint 2" )
ancom.all.filt.0.05.2 <- runAncom(ps2)
saveRDS(ancom.all.filt.0.05.2, file=paste0(output_data, "ANCOM/ancom_res2_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.2$out$detected_0.6)


#Timepoint 1
ps1<-subset_samples(ps_no_norm_filt, Within.study.sampling.date == "Timepoint 2" )
ancom.all.filt.0.05.1 <- runAncom(ps1)
saveRDS(ancom.all.filt.0.05.1, file=paste0(output_data, "ANCOM/ancom_res1_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.1$out$detected_0.6)

```

# Visualizing results from DESeq, MetagenomeSeq, and Ancom
### Read in results from above tests
```{r}
deseq_res_P1 <- readRDS(paste0(output_data, "DESeq/deseq_res_P1_adjp0.05.rds"))
deseq_res_P2 <- readRDS(paste0(output_data, "DESeq/deseq_res_P2_adjp0.05.rds"))
deseq_res_P3 <- readRDS(paste0(output_data, "DESeq/deseq_res_P2_adjp0.05.rds"))
deseq_res_time <- readRDS(paste0(output_data, "DESeq/Deseq_time_interaction_adjp0.05.rds"))

zig_res_P1 <- readRDS(paste0(output_data, "metagenomSeq/zig_res_P1_adjp0.05.rds"))
zig_res_P2 <- readRDS(paste0(output_data, "metagenomSeq/zig_res_P2_adjp0.05.rds"))
zig_res_P3 <- readRDS(paste0(output_data, "metagenomSeq/zig_res_P3_adjp0.05.rds"))
zig_res_time <- readRDS(paste0(output_data, "metagenomSeq/zig_res_time_adjp0.05.rds"))

#ANCOM results were all negative

#Load for plotting purposes:
ps_tss <- readRDS(paste0(output_data, "Normalized/ps_tss_pass_min_postDD_min0.03.rds"))
```

```{r visualization_fun}
### functions to plot
make_vis_plots <- function(ps_norm, grouping, tax, plot_type=c('box', 'bar')){
  # ps should be a normalized (DESeq or CSS) phyloseq object
  # grouping should match the column name in the sample_data
  # tax is a taxonomical bin id (ASV) in the counts table to plot
  
  # subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_norm) %in% tax, ps_norm)
  
  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')

  # boxplot
  if(plot_type=='box'){
    ggplot(data=plot_table, aes(x=Within.study.sampling.date, y = value, fill=Group)) + 
      geom_boxplot(outlier.color=NA) +
      geom_jitter(position=position_jitterdodge(0.2), cex=1.5, color="gray44") + 
      labs(title =deparse(substitute(ps_norm)), x='', y ='Proportional counts, log scale') + 
      scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  } else if (plot_type=='bar'){
    plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title =deparse(substitute(ps_norm)), x='', y ='Proportional counts, 0 to 1 scale') + 
      #scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  }
}

make_vis_plots_print <- function(ps_norm, grouping, tax, plot_type=c('box', 'bar'), plot_title){
  # ps should be a normalized (DESeq or CSS) phyloseq object
  # grouping should match the column name in the sample_data
  # tax is a taxonomical bin id (ASV) in the counts table to plot
  
  # subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_norm) %in% tax, ps_norm)
  
  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')

  # boxplot
  if(plot_type=='box'){
    ggplot(data=plot_table, aes(x=Within.study.sampling.date, y = value, fill=Group)) + 
      geom_boxplot(outlier.color=NA) +
      geom_jitter(position=position_jitterdodge(0.2), cex=1.5, color="gray44") + 
      labs(title =deparse(substitute(ps_norm)), x='', y ='Proportional counts, log scale') + 
      scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  } else if (plot_type=='bar'){
    plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title =plot_title, x='', y ='Proportional counts, 0 to 1 scale') + 
      #scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  }
}

```

```{r visualization_sig}
######BOXPLOT of significant ones
# make significant taxa into one table so that all pvalues retained
significant_tax=NULL
significant_tax <- merge(data.table(deseq_res_P1[[2]], keep.rownames=TRUE)[, list(rn, deseq_P1_adjp=padj)],
                         data.table(deseq_res_P2[[2]], keep.rownames=TRUE)[, list(rn, deseq_P2_adjp=padj)],
                         by='rn', all=TRUE)
significant_tax <- merge(significant_tax,
                         data.table(deseq_res_P3[[2]], keep.rownames=TRUE)[, list(rn, deseq_P3_adjp=padj)],
                         by='rn', all=TRUE)
significant_tax <- merge(significant_tax,
                         data.table(deseq_res_time[[2]], keep.rownames=TRUE)[, list(rn, deseq_timeseries_adjp=padj)],
                         by='rn', all=TRUE)
significant_tax <- merge(significant_tax,
                         data.table(zig_res_P1[[2]], keep.rownames=TRUE)[, list(rn, mtgseq_P1_adjp=adjPvalues)],
                         by='rn', all=TRUE)
significant_tax <- merge(significant_tax,
                         data.table(zig_res_P2[[2]], keep.rownames=TRUE)[, list(rn, mtgseq_P2_adjp=adjPvalues)],
                         by='rn', all=TRUE)
significant_tax <- merge(significant_tax,
                         data.table(zig_res_P3[[2]], keep.rownames=TRUE)[, list(rn, mtgseq_P3_adjp=adjPvalues)],
                         by='rn', all=TRUE)
significant_tax <- merge(significant_tax,
                         data.table(zig_res_time[[2]], keep.rownames=TRUE)[, list(rn, mtgseq_timeseries_adjp=adjPvalues)],
                         by='rn', all=TRUE)

# remove nothing
significant_tax <- significant_tax[rn!='nothing']

# write results
write.csv(significant_tax, file=paste0(output_data, 'Significant_res_deseq_q', deseq_cut, '_mtgseq_q', mtgseq_cut, '.csv'), row.names=FALSE)
          
datatable(significant_tax)

# also, find taxonomical annotations
# NOTE: single ASV may have multiple annotations due to tie hits
#Changing var all_tax_data to tax_table since I don't have this object since I don't have all_tax_data as a object,
df_res <- tax_table(ps_no_norm_filt)[rownames(tax_table(ps_no_norm_filt)) %in% significant_tax$rn]
```

### DESeq results
```{r visualization_sig_deseq, fig.width=12}
## plot
# common by deseq
com_deseq_taxa=significant_tax[!is.na(deseq_P1_adjp) & !is.na(deseq_P2_adjp) & !is.na(deseq_P3_adjp)]

if(nrow(com_deseq_taxa)>0){
  print(make_vis_plots(ps_tss, 'phenotype', com_deseq_taxa$rn, 'box'))
} else {
  print('no common DESeq significant taxa')
}
```

### DESeq timeseries results
```{r visualization_sig_deseq_ts, fig.width=12, fig.height=9}
# deseq timeseries
if(nrow(significant_tax[!is.na(deseq_timeseries_adjp)])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(deseq_timeseries_adjp)]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(deseq_timeseries_adjp)]$rn, 'bar'))
} else {
  print('no DESeq timeseries significant taxa')
}
```

### metagenomeSeq results
```{r visualization_sig_mtgseq, fig.width=12, fig.height=3}
# common by metagenomeseq
com_mtgseq_taxa=significant_tax[!is.na(mtgseq_P1_adjp) & !is.na(mtgseq_P2_adjp) & !is.na(mtgseq_P3_adjp)]

if(nrow(com_mtgseq_taxa)>0){
  print(make_vis_plots(ps_tss, 'phenotype', com_mtgseq_taxa$rn, 'box'))
} else {
  print('no common metagenomeSeq significant taxa')
}
```
### Meta_genome timeseries results
```{r visualization_sig_mtg_ts, fig.width=12, fig.height=9}
# mtgseq timeseries
if(nrow(significant_tax[!is.na(mtgseq_timeseries_adjp)])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[1:6], 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[1:6], 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}

if(nrow(significant_tax[!is.na(mtgseq_timeseries_adjp)])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[7:12], 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[7:12], 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}

if(nrow(significant_tax[!is.na(mtgseq_timeseries_adjp)])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[13:18], 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[13:18], 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}

if(nrow(significant_tax[!is.na(mtgseq_timeseries_adjp)])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[19:24], 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[19:24], 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}

if(nrow(significant_tax[!is.na(mtgseq_timeseries_adjp)])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[25:27], 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[!is.na(mtgseq_timeseries_adjp)]$rn[25:27], 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}
```
