---
title: "Make_Enrichment_Table_with_Associated_Variables_Labeled"
author: "Austin Martin"
date: "9/2/2020"
output: html_document
---

```{r table}

#Deseq results
ps_no_norm_filt <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/Filtered/ps_no_norm_friedfilt.rds")
deseq_res_P1 <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/DESeq/deseq_res_P1_adjp0.05.rds")
deseq_res_P2 <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/DESeq/deseq_res_P2_adjp0.05.rds")
deseq_res_P3 <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/DESeq/deseq_res_P3_adjp0.05.rds")

deseq_res_time <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/DESeq/Deseq_time_interaction_adjp0.05.rds")


#Add enrichment column
deseq_res_P1[[2]]$enrichment <- ifelse(deseq_res_P1[[2]]$log2FoldChange < 0, "Aut", "Control")
deseq_res_P2[[2]]$enrichment <- ifelse(deseq_res_P2[[2]]$log2FoldChange < 0, "Aut", "Control")
deseq_res_P3[[2]]$enrichment <- ifelse(deseq_res_P3[[2]]$log2FoldChange < 0, "Aut", "Control")

deseq_res_time[[2]]$enrichment <- ifelse(deseq_res_time[[2]]$log2FoldChange < 0, "Aut", "Control")


#Assemble Deseq results into one table
DesP1<-tibble(rownames(deseq_res_P1[[2]]), rep("DESeq2_P1", length(rownames(deseq_res_P1[[2]]))), deseq_res_P1[[2]]$enrichment)
colnames(DesP1) <- c("ASV", "Method+Data","Enrichment")

DesP2<-tibble(rownames(deseq_res_P2[[2]]), rep("DESeq2_P2", length(rownames(deseq_res_P2[[2]]))),deseq_res_P2[[2]]$enrichment)
colnames(DesP2) <- c("ASV", "Method+Data","Enrichment")

DesP3<-tibble(rownames(deseq_res_P3[[2]]), rep("DESeq2_P3", length(rownames(deseq_res_P3[[2]]))),deseq_res_P3[[2]]$enrichment)
colnames(DesP3) <- c("ASV", "Method+Data","Enrichment")

DesAll<-tibble(rownames(deseq_res_time[[2]]), rep("DESeq2_All", length(rownames(deseq_res_time[[2]]))),deseq_res_time[[2]]$enrichment)
colnames(DesAll) <- c("ASV", "Method+Data","Enrichment")

deseq_allsig<-rbind(DesP1,DesP2, DesP3, DesAll)

#Find ones found multiple times (either in indiviudual timepoints or all together) and add them back in with correct labeling

deseqcount<-plyr::ddply(deseq_allsig, "ASV", transform, count = length(ASV))
duplic<-tibble(deseqcount$ASV[which(deseqcount$count >= 2)], deseqcount$Method.Data[which(deseqcount$count >= 2)], deseqcount$Enrichment[which(deseqcount$count >= 2)])

replace <-paste(duplic$`deseqcount$Method.Data[which(deseqcount$count >= 2)]`[which(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]` ==  unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`)[1])], collapse = "")

replace_enrich<-paste(duplic$`deseqcount$Enrichment[which(deseqcount$count >= 2)]`[which(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]` ==  unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`)[1])], collapse = "")

tmp<-tibble(unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`)[1], replace, replace_enrich)
colnames(tmp) <- c("ASV", "Method+Data", "Enrichment")
replacetab <-tmp

for (i in 2:length(unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`))) {
  replace <-paste(duplic$`deseqcount$Method.Data[which(deseqcount$count >= 2)]`[which(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]` ==  unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`)[i])], collapse = "")
  replace_enrich<-paste(duplic$`deseqcount$Enrichment[which(deseqcount$count >= 2)]`[which(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]` ==  unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`)[i])], collapse = "")
  tmp<-tibble(unique(duplic$`deseqcount$ASV[which(deseqcount$count >= 2)]`)[i], replace, replace_enrich)
  colnames(tmp) <- c("ASV", "Method+Data", "Enrichment")
  replacetab<-rbind(replacetab, tmp)
}




deseq_allsig<-deseq_allsig[-which(deseq_allsig$ASV %in% replacetab$ASV),]

deseq_allsig<-rbind(deseq_allsig, replacetab)
deseq_allsig<-cbind(deseq_allsig, as(tax_table(ps_no_norm_filt)[deseq_allsig$ASV, ], "matrix"))
#saveRDS(deseq_allsig, "DESeq/full_res_table_deseq")

deseq_allsig.print <- deseq_allsig
deseq_allsig.print$ASV <- NULL
rownames(deseq_allsig.print) <- NULL
#deseq_allsig.print


#MTG results 

zig_res_P1 <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/metagenomseq/zig_res_P1_adjp0.05.rds")
zig_res_P2 <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/metagenomseq/zig_res_P2_adjp0.05.rds")
zig_res_P3 <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/metagenomseq/zig_res_P3_adjp0.05.rds")

zig_res_time <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/metagenomseq/zig_res_time_adjp0.05.rds")



# print significant taxa and add enrichment column
#datatable(zig_res_P1[[2]])
#datatable(zig_res_P2[[2]])
#datatable(zig_res_P3[[2]])
#datatable(zig_res_time[[2]])

zig_res_P1_filtered <- data.frame(cbind(zig_res_P1[[2]], tax_table(P1)[rownames(zig_res_P1[[2]]),]))
zig_res_P1_filtered$enriched <- ifelse(zig_res_P1_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_P2_filtered <- data.frame(cbind(zig_res_P2[[2]], tax_table(P2)[rownames(zig_res_P2[[2]]), ]))
zig_res_P2_filtered$enriched <- ifelse(zig_res_P2_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_P3_filtered <- data.frame(cbind(zig_res_P3[[2]], tax_table(P3)[rownames(zig_res_P3[[2]]), ]))
zig_res_P3_filtered$enriched <- ifelse(zig_res_P3_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_time_filtered <- data.frame(cbind(zig_res_time[[2]], tax_table(ps_no_norm_filt)[rownames(zig_res_time[[2]]), ]))
zig_res_time_filtered$enriched <- ifelse(zig_res_time_filtered$phenotypeN < 0, "Aut", "Control")


zig_res_P1[[2]]$enriched <-zig_res_P1_filtered$enriched
zig_res_P2[[2]]$enriched <-zig_res_P2_filtered$enriched
zig_res_P3[[2]]$enriched <-zig_res_P3_filtered$enriched
zig_res_time[[2]]$enriched <-zig_res_time_filtered$enriched


#Conglomerate
MtgP1<-tibble(rownames(zig_res_P1[[2]]), rep("Mtg_P1", length(rownames(zig_res_P1[[2]]))), zig_res_P1[[2]]$enriched)
colnames(MtgP1) <- c("ASV", "Method+Data", "Enrichment")

MtgP2<-tibble(rownames(zig_res_P2[[2]]), rep("Mtg_P2", length(rownames(zig_res_P2[[2]]))),zig_res_P2[[2]]$enriched)
colnames(MtgP2) <- c("ASV", "Method+Data", "Enrichment")

MtgP3<-tibble(rownames(zig_res_P3[[2]]), rep("Mtg_P3", length(rownames(zig_res_P3[[2]]))),zig_res_P3[[2]]$enriched)
colnames(MtgP3) <- c("ASV", "Method+Data", "Enrichment")

MtgAll<-tibble(rownames(zig_res_time[[2]]), rep("Mtg_P3", length(rownames(zig_res_time[[2]]))),zig_res_time[[2]]$enriched)
colnames(MtgAll) <- c("ASV", "Method+Data", "Enrichment")

Mtg_allsig<-rbind(MtgP1,MtgP2, MtgP3, MtgAll)

Mtgcount<-plyr::ddply(Mtg_allsig, "ASV", transform, count = length(ASV))
duplic<-tibble(Mtgcount$ASV[which(Mtgcount$count >= 2)], Mtgcount$Method.Data[which(Mtgcount$count >= 2)], Mtgcount$Enrichment[which(Mtgcount$count >= 2)])


replace <-paste(duplic$`Mtgcount$Method.Data[which(Mtgcount$count >= 2)]`[which(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]` ==  unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`)[1])], collapse = "")

 replace_enrich<-paste(duplic$`Mtgcount$Enrichment[which(Mtgcount$count >= 2)]`[which(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]` ==  unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`)[1])], collapse = "")

tmp<-tibble(unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`)[1], replace, replace_enrich)
colnames(tmp) <- c("ASV", "Method+Data", "Enrichment")
replacetab <-tmp

for (i in 2:length(unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`))) {
  replace <-paste(duplic$`Mtgcount$Method.Data[which(Mtgcount$count >= 2)]`[which(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]` ==  unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`)[i])], collapse = "")
  replace_enrich<-paste(duplic$`Mtgcount$Enrichment[which(Mtgcount$count >= 2)]`[which(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]` ==  unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`)[i])], collapse = "")

  tmp<-tibble(unique(duplic$`Mtgcount$ASV[which(Mtgcount$count >= 2)]`)[i], replace, replace_enrich)
  colnames(tmp) <- c("ASV", "Method+Data", "Enrichment")
  replacetab<-rbind(replacetab, tmp)
}

Mtg_allsig<-Mtg_allsig[-which(Mtg_allsig$ASV %in% replacetab$ASV),]
Mtg_allsig<-rbind(Mtg_allsig, replacetab)

Mtg_allsig<-cbind(Mtg_allsig, as(tax_table(ps_no_norm_filt)[Mtg_allsig$ASV, ], "matrix"))
#saveRDS(Mtg_allsig, "metagenomseq//full_res_table_mtg")

Mtg_allsig.print <- Mtg_allsig
Mtg_allsig.print$ASV <- NULL
rownames(Mtg_allsig.print) <- NULL
Mtg_allsig.print

#will concatenate mtg and des into one table

fullsigtab_esv<-rbind(Mtg_allsig, deseq_allsig)
fullsigtab_esv<-data.table(fullsigtab_esv)

fullcount<-plyr::ddply(.data = fullsigtab_esv, "ASV", transform, count =length(ASV))
duplic<-tibble(fullcount$ASV[which(fullcount$count >= 2)], fullcount$Method.Data[which(fullcount$count >= 2)], fullcount$Enrichment[which(fullcount$count >= 2)])


replace <-paste(duplic$`fullcount$Method.Data[which(fullcount$count >= 2)]`[which(duplic$`fullcount$ASV[which(fullcount$count >= 2)]` ==  unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`)[1])], collapse = "")

 replace_enrich<-paste(duplic$`fullcount$Enrichment[which(fullcount$count >= 2)]`[which(duplic$`fullcount$ASV[which(fullcount$count >= 2)]` ==  unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`)[1])], collapse = "")

tmp<-tibble(unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`)[1], replace, replace_enrich)
colnames(tmp) <- c("ASV", "Method+Data", "Enrichment")
replacetab <-tmp

for (i in 2:length(unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`))) {
  replace <-paste(duplic$`fullcount$Method.Data[which(fullcount$count >= 2)]`[which(duplic$`fullcount$ASV[which(fullcount$count >= 2)]` ==  unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`)[i])], collapse = "")
  replace_enrich<-paste(duplic$`fullcount$Enrichment[which(fullcount$count >= 2)]`[which(duplic$`fullcount$ASV[which(fullcount$count >= 2)]` ==  unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`)[i])], collapse = "")

  tmp<-tibble(unique(duplic$`fullcount$ASV[which(fullcount$count >= 2)]`)[i], replace, replace_enrich)
  colnames(tmp) <- c("ASV", "Method+Data", "Enrichment")
  replacetab<-rbind(replacetab, tmp)
}

fullsigtab_esv<-fullsigtab_esv[-which(fullsigtab_esv$ASV %in% replacetab$ASV),]
replacetab<-cbind(replacetab, as(tax_table(ps_no_norm_filt)[replacetab$ASV, ], "matrix"))

fullsigtab_esv<-rbind(fullsigtab_esv, replacetab)

#this one is odd, will label accordingly in Enrichment_Cat
fullsigtab_esv$Genus[-which(fullsigtab_esv$Enrichment == "ControlAutAut")]

fullsigtab_esv$Enrichment_Cat <- gsub("Aut", "x", fullsigtab_esv$Enrichment)
fullsigtab_esv$Enrichment_Cat <- gsub("xAut", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xAut", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xAut", "x", fullsigtab_esv$Enrichment_Cat)

fullsigtab_esv$Enrichment_Cat <- gsub("xxxxx", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xxxx", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xxx", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xx", "x", fullsigtab_esv$Enrichment_Cat)

fullsigtab_esv$Enrichment_Cat <- gsub("x", "Aut", fullsigtab_esv$Enrichment_Cat)

fullsigtab_esv$Enrichment_Cat <- gsub("Control", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xControl", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xControl", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xControl", "x", fullsigtab_esv$Enrichment_Cat)

fullsigtab_esv$Enrichment_Cat <- gsub("xxxxx", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xxxx", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xxx", "x", fullsigtab_esv$Enrichment_Cat)
fullsigtab_esv$Enrichment_Cat <- gsub("xx", "x", fullsigtab_esv$Enrichment_Cat)

fullsigtab_esv$Enrichment_Cat <- gsub("x", "Control", fullsigtab_esv$Enrichment_Cat)

fullsigtab_esv$Enrichment_Cat <- gsub("ControlAut", "Both_depending_on_timepoint", fullsigtab_esv$Enrichment_Cat)


#ANCOM

ancom_time<-readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/ANCOM/ancom_res0.05.rds")
ancom1<-readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/ANCOM/ancom_res1_0.05.rds")
ancom2<-readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/ANCOM/ancom_res2_0.05.rds")
ancom3<-readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/ANCOM/ancom_res3_0.05.rds")

fullsigtab_esv$ANCOM_sig <- rep("FALSE", length(fullsigtab_esv$ASV))

fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom_time$out$taxa_id[which(ancom_time$out$detected_0.6 == TRUE)])] <- rep("TRUE", length(fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom_time$out$taxa_id[which(ancom_time$out$detected_0.6 == TRUE)])]))

fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom1$out$taxa_id[which(ancom1$out$detected_0.6 == TRUE)])] <- rep("TRUE", length(fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom1$out$taxa_id[which(ancom1$out$detected_0.6 == TRUE)])]))

fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom2$out$taxa_id[which(ancom2$out$detected_0.6 == TRUE)])] <- rep("TRUE", length(fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom2$out$taxa_id[which(ancom2$out$detected_0.6 == TRUE)])]))

fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom3$out$taxa_id[which(ancom3$out$detected_0.6 == TRUE)])] <- rep("TRUE", length(fullsigtab_esv$ANCOM_sig[which(fullsigtab_esv$ASV %in% ancom3$out$taxa_id[which(ancom3$out$detected_0.6 == TRUE)])]))

#only ancom sig esv not found in fullsigtab (individual time point ancom sig esvs were found in sigtab, hence only leaving code for ancom_time)
new_ancom_esv<-ancom_time$out$taxa_id[which(ancom_time$out$detected_0.6 == TRUE)][-which(ancom_time$out$taxa_id[which(ancom_time$out$detected_0.6 == TRUE)] %in% fullsigtab_esv$ASV)]

new_ancom_tab<-tibble(new_ancom_esv, "ANCOM", "Enrichment")
colnames(new_ancom_tab) <- c("ASV", "Method+Data", "Enrichment")
new_ancom_tab<-cbind(new_ancom_tab, as(tax_table(ps_no_norm_filt)[as.character(new_ancom_esv), ], "matrix"))


#Find Enrichment for it
### functions to plot (WORKS ONLY IN R3.6.2)
make_vis_plots_print <- function(ps_norm, grouping, tax, plot_type=c('box', 'bar'), plot_title){
  # ps should be a normalized (DESeq or CSS) phyloseq object
  # grouping should match the column name in the sample_data
  # tax is a taxonomical bin id (ASV) in the counts table to plot
  
  # subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_norm) %in% tax, ps_norm)
  
  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')

  # boxplot
  if(plot_type=='box'){
    ggplot(data=plot_table, aes(x=Within.study.sampling.date, y = value, fill=Group)) + 
      geom_boxplot(outlier.color=NA) +
      geom_jitter(position=position_jitterdodge(0.2), cex=1.5, color="gray44") + 
      labs(title =deparse(substitute(ps_norm)), x='', y ='Proportional counts, log scale') + 
      scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  } else if (plot_type=='bar'){
    plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title =plot_title, x='', y ='Proportional counts, 0 to 1 scale') + 
      #scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  }
}


ps <- readRDS("~/M3_Datasets_updated/M3_Phyloseq_Analysis/results/Normalized/ps_tss_pass_min_postDD_min0.03.rds")
make_vis_plots_print(ps, 'phenotype', as.character(new_ancom_esv), 'bar', new_ancom_tab$Genus)

#Def in enriched in control

new_ancom_tab$Enrichment <- "Control"
new_ancom_tab$ANCOM_sig <- "TRUE"
new_ancom_tab$Enrichment_Cat <- "Control"


#combine into fulltab

fullsigtab_esv<-rbind(fullsigtab_esv, new_ancom_tab)


write.csv(fullsigtab_esv, file = "~/M3_Datasets_updated/M3_Phyloseq_Analysis/EnrichmentASV_Table_with_ANCOM_9_4_2020.csv")

```


