---
title: "plotting_taxa_abundance"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(reshape2)
library(phyloseq)
library(DT)
library(data.table)

sgColorPalette = c("#84CF04","#01B5BB","#E50E63","#6D7272","#8F389E",
                     "#DF8236","#036B6B","#F1BA2F","#9F832D","#94E804",
                     "#01D4DB","#FAC131","#B0B8B8","#F08C3A","#FF106E",
                     "#B948CC","#05B5B5","#CFAA3A")
```

# Visualizing results from DESeq, MetagenomeSeq, and Ancom
### Read in results from above tests
```{r}
output_data <- "results/"
deseq_res_P1 <- readRDS(paste0(output_data, "DESeq/deseq_res_P1_adjp0.05.rds"))
deseq_res_P2 <- readRDS(paste0(output_data, "DESeq/deseq_res_P2_adjp0.05.rds"))
deseq_res_P3 <- readRDS(paste0(output_data, "DESeq/deseq_res_P2_adjp0.05.rds"))
deseq_res_time <- readRDS(paste0(output_data, "DESeq/Deseq_time_interaction_adjp0.05.rds"))

zig_res_P1 <- readRDS(paste0(output_data, "metagenomSeq/zig_res_P1_adjp0.05.rds"))
zig_res_P2 <- readRDS(paste0(output_data, "metagenomSeq/zig_res_P2_adjp0.05.rds"))
zig_res_P3 <- readRDS(paste0(output_data, "metagenomSeq/zig_res_P3_adjp0.05.rds"))
zig_res_time <- readRDS(paste0(output_data, "metagenomSeq/zig_res_time_adjp0.05.rds"))

#ANCOM results were all negative


significant_tax <- read.csv(paste0(output_data, 'Significant_res_deseq_q0.05_mtgseq_q0.05_tax.csv'))
datatable(significant_tax)

#Load for plotting purposes:
ps_tss <- readRDS(paste0(output_data, "Normalized/ps_tss_pass_min_postDD_min0.03.rds"))

```


```{r visualization_fun}
### functions to plot
make_vis_plots <- function(ps_norm, grouping, tax, plot_type=c('box', 'bar')){
  # ps should be a normalized (DESeq or CSS) phyloseq object
  # grouping should match the column name in the sample_data
  # tax is a taxonomical bin id (ASV) in the counts table to plot
  
  # subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_norm) %in% tax, ps_norm)

  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  #taxa names for plot labels
  species_names <- as.character(ps_tss@tax_table[plot_table$rn, 7])
  genus_names <- as.character(ps_tss@tax_table[plot_table$rn, 6])
  unclassified <- species_names == "s__unclassified"
  species_names[unclassified] <- genus_names[unclassified] 
  plot_table$rn <- species_names
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')

  # boxplot
  if(plot_type=='box'){
    ggplot(data=plot_table, aes(x=Within.study.sampling.date, y = value, fill=Group)) + 
      geom_boxplot(outlier.color=NA) +
      geom_jitter(position=position_jitterdodge(0.2), cex=1.5, color="gray44") + 
      labs(title =deparse(substitute(ps_norm)), x='', y ='Proportional counts, log scale') + 
      scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  } else if (plot_type=='bar'){
    plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title =deparse(substitute(ps_norm)), x='', y ='Proportional counts, 0 to 1 scale') +
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  }
}


```

## Both DESeq and metagenomeseq timeseries
```{r fig.width=12, fig.height=10}
keep <- !is.na(significant_tax$deseq_timeseries_adjp) & !is.na(significant_tax$mtgseq_timeseries_adjp)
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep, ]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep, ]$rn, 'bar'))
} else {
  print('no combined timeseries significant taxa')
}
```

### At least 1 timepoint in both DESeq and metagenomeSeq
```{r fig.width=12, fig.height=7}
keep_deseq <- !is.na(significant_tax$deseq_P1_adjp) | !is.na(significant_tax$deseq_P2_adjp) | !is.na(significant_tax$deseq_P3_adjp)
keep_metagenome <- !is.na(significant_tax$mtgseq_P1_adjp) | !is.na(significant_tax$mtgseq_P2_adjp) | !is.na(significant_tax$mtgseq_P3_adjp)
keep <- keep_deseq & keep_metagenome
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep, ]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep, ]$rn, 'bar'))
} else {
  print('no DESeq timeseries significant taxa')
}
```


### DESeq all 3 timepoints
```{r visualization_sig_deseq, fig.width=12, fig.height=4}
## plot
# common by deseq
com_deseq_taxa = significant_tax[!is.na(significant_tax$deseq_P1_adjp) & !is.na(significant_tax$deseq_P2_adjp) & !is.na(significant_tax$deseq_P3_adjp), ]

if(nrow(com_deseq_taxa)>0){
  print(make_vis_plots(ps_tss, 'phenotype', com_deseq_taxa$rn, 'box'))
  print(make_vis_plots(ps_tss, 'phenotype', com_deseq_taxa$rn, 'bar'))
} else {
  print('no common DESeq significant taxa')
}
```

### metagenomeSeq all 3 timepoints
```{r visualization_sig_mtgseq, fig.width=12, fig.height=10}
# common by metagenomeseq
com_mtgseq_taxa=significant_tax[!is.na(significant_tax$mtgseq_P1_adjp) & !is.na(significant_tax$mtgseq_P2_adjp) & !is.na(significant_tax$mtgseq_P3_adjp), ]

if(nrow(com_mtgseq_taxa)>0){
  print(make_vis_plots(ps_tss, 'phenotype', com_mtgseq_taxa$rn, 'box'))
} else {
  print('no common metagenomeSeq significant taxa')
}
```


### DESeq at least 2 timepoints
```{r visualization_sig_deseq_ts, fig.width=12, fig.height=10}
# deseq timeseries
keep <- data.frame(!is.na(significant_tax$deseq_P1_adjp), !is.na(significant_tax$deseq_P2_adjp), !is.na(significant_tax$deseq_P3_adjp))
keep <- apply(keep, 1, function(x) sum(x)>1)
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep, ]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep, ]$rn, 'bar'))
} else {
  print('no DESeq timeseries significant taxa')
}
```



### Meta_genome at least 2 timepoints
```{r visualization_sig_mtg_ts, fig.width=12, fig.height=5}
keep <- data.frame(!is.na(significant_tax$mtgseq_P1_adjp), !is.na(significant_tax$mtgseq_P2_adjp), !is.na(significant_tax$mtgseq_P3_adjp))
keep <- apply(keep, 1, function(x) sum(x)>1)
# mtgseq timeseries
if(nrow(significant_tax[!is.na(significant_tax$mtgseq_timeseries_adjp),])>0){
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep,]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', significant_tax[keep,]$rn, 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}

```