---
title: "16s_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(reshape2)
library(phyloseq)
library(rstatix)
library(ggpubr)
library(pbapply)
library(DESeq2)
library(DT)
library(metagenomeSeq)
library(data.table)
library(tibble)
library(exactRankTests)
source("~/ANCOM2.1/ANCOM/scripts/ancom_v2.1.R")

```

### read phyloseq

```{r}
output_data = "../results/"
ps_deseq <- readRDS(paste0(output_data, "Filtered/ps_deseq_friedfilt.rds"))
ps_css <- readRDS(paste0(output_data,"Filtered/ps_css_friedfilt.rds"))
ps_no_norm_filt <- readRDS(paste0(output_data,"Filtered/ps_no_norm_friedfilt.rds"))

ps_no_norm_filt_name_edit <- ps_no_norm_filt 
#adjust genus names for a proper conglomeration
# fix taxonomic assignments
strain_assignment <- read.csv("../data/INRD17_0415_combinedRuns_multi_hits.csv")
strain_assignment <- strain_assignment[!duplicated(strain_assignment$ss_id), ]
rownames(strain_assignment) <- strain_assignment$ss_id
strain_names <- taxa_names(ps_no_norm_filt_name_edit)[grepl("t__", taxa_names(ps_no_norm_filt_name_edit))]
names_split <- strsplit(as.character(strain_assignment[strain_names, ]$display_name), split = ".", fixed = T)
genus <- unlist(lapply(names_split, function(x) return(x[[1]])))
species <- unlist(lapply(names_split, function(x) return(x[[2]])))
ps_no_norm_filt_name_edit@tax_table[strain_names, 6] <- paste("g__", genus, sep = "")
#ps_no_norm_filt_genus_conglom@tax_table[strain_names, 7] <- paste("s__", species, sep = "")

tax <-as.data.frame(ps_no_norm_filt_name_edit@tax_table)
#tax<-tax2
#change PROV placeholder to be consistent with dada2 name assignments

tax$Genus[grep(pattern = "PROV", tax$Genus)] <- "g__unclassified"
tax$Family[grep(pattern = "PROV", tax$Family)] <- "f__unclassified"

#change other name differences
tax$Genus <- gsub(pattern = "g__", "", tax$Genus)
tax$Genus <- gsub(pattern = "_.", "", tax$Genus)

tax$Genus[tax$Genus == "clostridiaceae"] <- "clostridium"

tax$Genus <- tolower(tax$Genus)

#remove NA species/strain columns in conglomerated phyloseq to avoid metagenomeseq error so only columns 1-6
ps_no_norm_filt_name_edit@tax_table <- tax_table(as.matrix(tax[, 1:6]))

ps_no_norm_filt_genus_conglom <- tax_glom(ps_no_norm_filt_name_edit, "Genus")

```

### Cutoff parameters
```{r cutoff}
# min post DADA2 counts to run analysis
min_postDD=20000
# DESeq significant cutoff
deseq_cut=0.05
# metagenomeSeq significant cutoff
mtgseq_cut=0.05
# PERMANOVA pvalue and R2 cutoff for visualization
permanova_pcut=0.05
permanova_cut=0.1
```

### DESeq
```{r}
###Run DESeq proper (not just the normalization but all of it)
dir.create(paste0(output_data, 'DESeq/'))
runDESeq <- function(ps, dcut, time = FALSE){
  if(time == TRUE){
    #Working with time series
    #according to the DeSeq vignette: design including the time factor, and then test using the likelihood ratio test as described
    #the following section, where the time factor is removed in the reduced formula
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype + Within.study.sampling.date) 
  }else{
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype) 
  }
  
  #diagdds <- estimateSizeFactors(diagdds, type = "poscounts")
  diagdds <- DESeq(diagdds,fitType="parametric", betaPrior = FALSE) 
  res = results(diagdds, contrast = c("phenotype", "N", "A"))
  res$padj[is.na(res$padj)] = 1
  sig <- res[res$padj < dcut,]
  if (dim(sig)[1] == 0) 
  {sigtab<- as.data.frame(1, row.names="nothing")
  colnames(sigtab) <- 'padj'}
  else 
  {
    sigtab <- data.frame(sig)
  }
  return(list(res, sigtab))
}


###Running analysis 
###split thedata based on the real 3 timepoints
P1<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 1"], ps_no_norm_filt)
P2<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 2"], ps_no_norm_filt)
P3<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 3"], ps_no_norm_filt)

#several significants 
deseq_res_P1 <- runDESeq(P1, deseq_cut) 
deseq_res_P2 <- runDESeq(P2, deseq_cut)
deseq_res_P3 <- runDESeq(P3, deseq_cut)

# print significant taxa
data.table(deseq_res_P1[[2]])
data.table(deseq_res_P2[[2]])
data.table(deseq_res_P3[[2]])
#"ASV_1669" present twice timepoint 1 and 3 


data.table(as.data.frame(ps_no_norm_filt@tax_table[intersect(rownames(deseq_res_P1[[2]]), rownames(deseq_res_P2[[2]])), ]))

# save
saveRDS(deseq_res_P1, file=paste0(output_data, "DESeq/deseq_res_P1_adjp", deseq_cut, ".rds"))
saveRDS(deseq_res_P2, file=paste0(output_data, "DESeq/deseq_res_P2_adjp", deseq_cut, ".rds"))
saveRDS(deseq_res_P3, file=paste0(output_data, "DESeq/deseq_res_P3_adjp", deseq_cut, ".rds"))



#Including timepoint in the formula
deseq_res_time <- runDESeq(ps_no_norm_filt, deseq_cut, time = TRUE)
data.table(as.data.frame(ps_no_norm_filt@tax_table[rownames(deseq_res_time[[2]]), ]))
saveRDS(deseq_res_time, file=paste0(output_data, "DESeq/Deseq_time_interaction_adjp", deseq_cut, ".rds"))
```

### MetagenomeSeq
```{r}
dir.create(paste0(output_data, 'metagenomseq/'))
###Run ZIG model fitting and prediction
run_metagenom_seq<-function(ps,maxit, mcut, time = F){
  p_metag<-phyloseq_to_metagenomeSeq(ps)
  #filtering at least 4 samples 
  p_metag= cumNorm(p_metag, p=0.75)
  normFactor =normFactors(p_metag)
  normFactor =log2(normFactor/median(normFactor) + 1)
  if(time == F){
    mod = model.matrix(~phenotype + normFactor, data = pData(p_metag))
  }
  if(time == T){
    mod = model.matrix(~phenotype + Within.study.sampling.date +normFactor, data = pData(p_metag))
  }
  
  settings =zigControl(maxit =maxit, verbose =FALSE)
  fit =fitZig(obj = p_metag, mod = mod, useCSSoffset = FALSE, control = settings)
  #Note: changed fit$taxa to fit@taxa in light of error (probably from newer metagenomeseq ver.)
  res_fit_nonfiltered <- MRtable(fit, number = length(fit@taxa))
  res_fit<-res_fit_nonfiltered[res_fit_nonfiltered$adjPvalues<mcut,]
  #finally remove the ones that are not with enough samples
  Min_effec_samp<-calculateEffectiveSamples(fit)
  Min_effec_samp<-Min_effec_samp[ names(Min_effec_samp)  %in% rownames(res_fit)] #####there is a bug here 
  #manually removing the ones with "NA"
  res_fit<-res_fit[grep("NA",rownames(res_fit), inv=T),]
  res_fit$Min_sample<-Min_effec_samp
  res_fit<-res_fit[res_fit$`+samples in group 0` >= Min_effec_samp & res_fit$`+samples in group 1` >= Min_effec_samp,]
  return(list(res_fit_nonfiltered, res_fit))
}
P1<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 1"], ps_no_norm_filt)
P2<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 2"], ps_no_norm_filt)
P3<-prune_samples(sample_names(ps_no_norm_filt)[ps_no_norm_filt@sam_data$Within.study.sampling.date == "Timepoint 3"], ps_no_norm_filt)

zig_res_P1 <- run_metagenom_seq(P1,30, mtgseq_cut) # 30The maximum number of iterations for the expectation-maximization algorithm
zig_res_P2 <- run_metagenom_seq(P2,30, mtgseq_cut)
zig_res_P3 <- run_metagenom_seq(P3,30, mtgseq_cut)

zig_res_time<- run_metagenom_seq(ps_no_norm_filt,30, mtgseq_cut, time = T)

# print significant taxa
datatable(zig_res_P1[[2]])
datatable(zig_res_P2[[2]])
datatable(zig_res_P3[[2]])
datatable(zig_res_time[[2]])

zig_res_P1_filtered <- data.frame(cbind(zig_res_P1[[2]], tax_table(P1)[rownames(zig_res_P1[[2]]),]))
zig_res_P1_filtered$enriched <- ifelse(zig_res_P1_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_P2_filtered <- data.frame(cbind(zig_res_P2[[2]], tax_table(P2)[rownames(zig_res_P2[[2]]), ]))
zig_res_P2_filtered$enriched <- ifelse(zig_res_P2_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_P3_filtered <- data.frame(cbind(zig_res_P3[[2]], tax_table(P3)[rownames(zig_res_P3[[2]]), ]))
zig_res_P3_filtered$enriched <- ifelse(zig_res_P3_filtered$phenotypeN < 0, "Aut", "Control")

zig_res_time_filtered <- data.frame(cbind(zig_res_time[[2]], tax_table(ps_no_norm_filt)[rownames(zig_res_time[[2]]), ]))
zig_res_time_filtered$enriched <- ifelse(zig_res_time_filtered$phenotypeN < 0, "Aut", "Control")


saveRDS(zig_res_P1, file=paste0(output_data, "metagenomseq/zig_res_P1_adjp", mtgseq_cut, ".rds"))
saveRDS(zig_res_P2, file=paste0(output_data, "metagenomseq/zig_res_P2_adjp", mtgseq_cut, ".rds"))
saveRDS(zig_res_P3, file=paste0(output_data, "metagenomseq/zig_res_P3_adjp", mtgseq_cut, ".rds"))
saveRDS(zig_res_time, file=paste0(output_data, "metagenomseq/zig_res_time_adjp", mtgseq_cut, ".rds"))

#do we have any in ESV in common?
#Reduce(intersect, list(rownames(zig_res_P1_filtered),rownames(zig_res_P2_filtered),rownames(zig_res_P3_filtered)))
#rownames(zig_res_P1[[2]])[which(rownames(zig_res_P1[[2]]) %in% c(rownames(zig_res_P2[[2]]), rownames(zig_res_P3[[2]])))]
#rownames(zig_res_P2[[2]])[which(rownames(zig_res_P2[[2]]) %in% rownames(zig_res_P3[[2]]))]

metag_res<-cbind(rownames(zig_res_time[[2]]), as(tax_table(ps_no_norm_filt)[rownames(zig_res_time[[2]]), ], "matrix"))
metag_res_esv <-metag_res
rownames(metag_res) <- NULL

```

### Ancom
```{r}

runAncom <- function(ps){
  metada_ps<-sample_data(ps)
  metada_ps<-as.data.frame(metada_ps)
  metada_ps$HostName <- as.character(metada_ps$Host.Name)
  metada_ps$phenotype <- as.character(metada_ps$phenotype)
  
  metada_ps<-tibble(metada_ps$HostName, metada_ps$phenotype, rows = rownames(metada_ps))
  colnames(metada_ps) <- c("HostName", "phenotype", "Biospecimen.Barcode")
  res_table_filt<-otu_table(ps)
  res_table_filt<-as.data.frame(res_table_filt)
  
  prepro<-feature_table_pre_process(feature_table = res_table_filt, meta_data = metada_ps, sample_var = "Biospecimen.Barcode", group_var = "phenotype", out_cut = 0.05, zero_cut = 0.90, lib_cut = 1000, neg_lb = FALSE)
  
  feature_table = prepro$feature_table # Preprocessed feature table
  meta_data = prepro$meta_data # Preprocessed metadata
  struc_zero = prepro$structure_zeros # Structural zero info
  
  main_var = "phenotype"; p_adj_method = "BH"; alpha = 0.05
  adj_formula = NULL; rand_formula = NULL
  ancom_res <- ANCOM(feature_table, meta_data, struc_zero, main_var, p_adj_method, 
                     alpha, adj_formula, rand_formula)#, control)
  return(ancom_res)
}

ancom.all.filt.0.05 <- runAncom(ps_no_norm_filt)
dir.create(path = "ANCOM")
saveRDS(ancom.all.filt.0.05, file=paste0(output_data, "ANCOM/ancom_res", mtgseq_cut, ".rds"))
#no significant taxa from ANCOM



#Analyze each timepoint independently
#Timepoint 3
ps3<-subset_samples(ps_no_norm_filt, Within.study.sampling.date == "Timepoint 3" )
ancom.all.filt.0.05.3 <- runAncom(ps3)
saveRDS(ancom.all.filt.0.05.3, file=paste0(output_data, "ANCOM/ancom_res3_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.3$out$detected_0.6)
# No sig


#Timepoint 2
ps2<-subset_samples(ps_no_norm_filt, Within.study.sampling.date == "Timepoint 2" )
ancom.all.filt.0.05.2 <- runAncom(ps2)
saveRDS(ancom.all.filt.0.05.2, file=paste0(output_data, "ANCOM/ancom_res2_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.2$out$detected_0.6)


#Timepoint 1
ps1<-subset_samples(ps_no_norm_filt, Within.study.sampling.date == "Timepoint 1" )
ancom.all.filt.0.05.1 <- runAncom(ps1)
saveRDS(ancom.all.filt.0.05.1, file=paste0(output_data, "ANCOM/ancom_res1_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.1$out$detected_0.6)

```

### Put together and save results
```{r save_sig_taxa}
output_data = "../results/"
mtgseq_cut = 0.05
deseq_cut = 0.05
return_dataframe <- function(df){
  tmp = data.frame(asv = rownames(df), padj = df$padj,
                   enrichment = ifelse(df$log2FoldChange < 0, "Aut", "Control"),
                   method = rep("DESeq2", nrow(df)))
  tmp <- tmp %>% mutate_all(as.character)
  return(tmp)
}

return_dataframe_zig <- function(df){
  tmp = data.frame(asv = rownames(df), padj = df$adjPvalues,
                   enrichment = ifelse(df$phenotypeN < 0, "Aut", "Control"),
                   method = rep("ZIG", nrow(df)))
  tmp <- tmp %>% mutate_all(as.character)
  return(tmp)
}

return_dataframe_ancom <- function(df){
  taxa <- as.character(df$taxa_id[df$detected_0.6 == T])
  print(length(taxa))
  if(length(taxa) > 0){
    ps_aut <- prune_samples(ps_css@sam_data$phenotype == 'A', ps_css)
    ps_nt <- prune_samples(ps_css@sam_data$phenotype == 'N', ps_css)
    #find directionality
    t_stat = t.test(as.numeric(ps_aut@otu_table[taxa, ]), as.numeric(ps_nt@otu_table[taxa, ]))$statistic
    enrichment <- ifelse(t_stat > 0, "Aut", "Control") # a t statistic that is positive means group1 is higher than group2. In our case group1 is au
    tmp <- data.frame(asv = taxa, padj = rep(NA, length(taxa)), enrichment = enrichment, method = rep("ANCOM", length(taxa)))
    tmp <- tmp %>% mutate_all(as.character)
    return(tmp)
  }else{
    return(rep(NA, 4))
  }

}
ps_css <- readRDS(paste0(output_data, "Normalized/ps_CSS_pass_min_postDD_min0.03.rds"))

#DESeq results
deseq_res_P1 <- readRDS(file=paste0(output_data, "DESeq/deseq_res_P1_adjp", deseq_cut, ".rds"))[[2]]
deseq_res_P1 <- return_dataframe(deseq_res_P1)
deseq_res_P1$method <- rep("DESEQ_P1", nrow(deseq_res_P1))

deseq_res_P2 <- readRDS(file=paste0(output_data, "DESeq/deseq_res_P2_adjp", deseq_cut, ".rds"))[[2]]
deseq_res_P2 <- return_dataframe(deseq_res_P2)
deseq_res_P2$method <- rep("DESEQ_P2", nrow(deseq_res_P2))

deseq_res_P3 <- readRDS( file=paste0(output_data, "DESeq/deseq_res_P3_adjp", deseq_cut, ".rds"))[[2]]
deseq_res_P3 <- return_dataframe(deseq_res_P3)
deseq_res_P3$method <- rep("DESEQ_P3", nrow(deseq_res_P3))

deseq_res_time <- readRDS( file=paste0(output_data, "DESeq/Deseq_time_interaction_adjp", deseq_cut, ".rds"))[[2]]
deseq_res_time <- return_dataframe(deseq_res_time)
deseq_res_time$method <- rep("DESEQ_time", nrow(deseq_res_time))

deseq_df <- rbind(deseq_res_P1, deseq_res_P2, deseq_res_P3, deseq_res_time)
deseq_df <- deseq_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))

#Metagenome_seq results
zig_res_P1 <- readRDS( file=paste0(output_data, "metagenomseq/zig_res_P1_adjp", mtgseq_cut, ".rds"))[[2]]
zig_res_P1 <- return_dataframe_zig(zig_res_P1)
zig_res_P1$method <- rep("ZIG_P1", nrow(zig_res_P1))

zig_res_P2 <- readRDS(file=paste0(output_data, "metagenomseq/zig_res_P2_adjp", mtgseq_cut, ".rds"))[[2]]
zig_res_P2 <- return_dataframe_zig(zig_res_P2)
zig_res_P2$method <- rep("ZIG_P2", nrow(zig_res_P2))

zig_res_P3 <- readRDS( file=paste0(output_data, "metagenomseq/zig_res_P3_adjp", mtgseq_cut, ".rds"))[[2]]
zig_res_P3 <- return_dataframe_zig(zig_res_P3)
zig_res_P3$method <- rep("ZIG_P3", nrow(zig_res_P3))

zig_res_time <- readRDS( file=paste0(output_data, "metagenomseq/zig_res_time_adjp", mtgseq_cut, ".rds"))[[2]]
zig_res_time <- return_dataframe_zig(zig_res_time)
zig_res_time$method <- rep("ZIG_time", nrow(zig_res_time))

zig_df <- rbind(zig_res_P1, zig_res_P2, zig_res_P3, zig_res_time)
zig_df <- zig_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))

#Ancom results

#Empty
#ancom_res_P1 <- readRDS(file=paste0(output_data, "ANCOM/ancom_res1_", mtgseq_cut, ".rds"))$out
#ancom_res_P1 <- return_dataframe_ancom(ancom_res_P1)
#ancom_res_P1$method <- rep("ANCOM_P1", nrow(ancom_res_P1))

#Empty
#ancom_res_P2 <- readRDS( file=paste0(output_data, "ANCOM/ancom_res2_", mtgseq_cut, ".rds"))$out
#ancom_res_P2 <- return_dataframe_ancom(ancom_res_P2)
#ancom_res_P2$method <- rep("ANCOM_P2", nrow(ancom_res_P2))

#Empty
#ancom_res_P3 <- readRDS( file=paste0(output_data, "ANCOM/ancom_res3_", mtgseq_cut, ".rds"))$out
#ancom_res_P3 <- return_dataframe_ancom(ancom_res_P3)
#ancom_res_P3$method <- rep("ancom_P3", nrow(ancom_res_P3))

ancom_res_time <- readRDS( file=paste0(output_data, "ANCOM/ancom_res", mtgseq_cut, ".rds"))$out
ancom_res_time <- return_dataframe_ancom(ancom_res_time)
ancom_res_time$method <- rep("ANCOM_time", nrow(ancom_res_time))

ancom_df <- ancom_res_time #ignore warning and it used to be rbind(ancom_time, p1, etc, but now only time has results)
ancom_df <- ancom_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))
keep <- apply(ancom_df, 1, function(x) return(!is.na(x[1])))
ancom_df <- ancom_df[keep, ]


# fix taxonomic assignments
strain_assignment <- read.csv("../data/INRD17_0415_combinedRuns_multi_hits.csv")
strain_assignment <- strain_assignment[!duplicated(strain_assignment$ss_id), ]
rownames(strain_assignment) <- strain_assignment$ss_id
strain_names <- taxa_names(ps_css)[grepl("t__", taxa_names(ps_css))]
names_split <- strsplit(as.character(strain_assignment[strain_names, ]$display_name), split = ".", fixed = T)
genus <- unlist(lapply(names_split, function(x) return(x[[1]])))
species <- unlist(lapply(names_split, function(x) return(x[[2]])))
ps_css@tax_table[strain_names, 6] <- paste("g__", genus, sep = "")
ps_css@tax_table[strain_names, 7] <- paste("s__", species, sep = "")

species_names <- as.character(ps_css@tax_table[, 7])
species_names <- unlist(lapply(strsplit(species_names, "__"), function(x) return(x[length(x)])))
species_names <- paste0("s__", species_names)
ps_css@tax_table[, 7] <- species_names


#all results
results_all <- rbind(deseq_df, zig_df, ancom_df)
results_all <- results_all %>% mutate_all(as.character)
results_grouped <- results_all %>% group_by(asv) %>% summarize(enrichment = paste0(enrichment, collapse = ", "), method = paste0(method, collapse = ", "))
results_grouped <- data.frame(cbind(results_grouped, ps_css@tax_table[results_grouped$asv, ]))
write.table(results_grouped, paste0(output_data, "significant_taxa_all.tsv"), row.names = F, sep = "\t")
data.table(results_grouped)

#time series
results_df <- rbind(deseq_res_time, zig_res_time, ancom_res_time)
results_df <- results_df %>% mutate_all(as.character)
results_df_grouped <- results_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))
#add taxa information
results_df_grouped <- data.frame(cbind(results_df_grouped, ps_css@tax_table[results_df_grouped$asv, ]))
keep <- apply(results_df_grouped, 1, function(x) return(all(x['enrichment'][[1]] == "Aut") | all(x['enrichment'][[1]] == "Control"))) 
results_final<- results_df_grouped[keep, ]
results_final$enrichment <- unlist(lapply(results_final$enrichment, function(x) return(paste0(x, collapse = ","))))
data.table(results_final)
write.table(results_final, paste0(output_data, "significant_taxa_timeseries.tsv"), row.names = F, sep = "\t")


```


### Repeat but with genus comglomerated 

#### DESeq
```{r}
###Run DESeq proper (not just the normalization but all of it)
dir.create(paste0(output_data, 'DESeq/'))
runDESeq <- function(ps, dcut, time = FALSE){
  if(time == TRUE){
    #Working with time series
    #according to the DeSeq vignette: design including the time factor, and then test using the likelihood ratio test as described
    #the following section, where the time factor is removed in the reduced formula
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype + Within.study.sampling.date) 
  }else{
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype) 
  }
  
  #diagdds <- estimateSizeFactors(diagdds, type = "poscounts")
  diagdds <- DESeq(diagdds,fitType="parametric", betaPrior = FALSE) 
  res = results(diagdds, contrast = c("phenotype", "N", "A"))
  res$padj[is.na(res$padj)] = 1
  sig <- res[res$padj < dcut,]
  if (dim(sig)[1] == 0) 
  {sigtab<- as.data.frame(1, row.names="nothing")
  colnames(sigtab) <- 'padj'}
  else 
  {
    sigtab <- data.frame(sig)
  }
  return(list(res, sigtab))
}


###Running analysis 
###split thedata based on the real 3 timepoints
P1<-prune_samples(sample_names(ps_no_norm_filt_genus_conglom)[ps_no_norm_filt_genus_conglom@sam_data$Within.study.sampling.date == "Timepoint 1"], ps_no_norm_filt_genus_conglom)
P2<-prune_samples(sample_names(ps_no_norm_filt_genus_conglom)[ps_no_norm_filt_genus_conglom@sam_data$Within.study.sampling.date == "Timepoint 2"], ps_no_norm_filt_genus_conglom)
P3<-prune_samples(sample_names(ps_no_norm_filt_genus_conglom)[ps_no_norm_filt_genus_conglom@sam_data$Within.study.sampling.date == "Timepoint 3"], ps_no_norm_filt_genus_conglom)

#several significants 
deseq_res_P1_genus <- runDESeq(P1, deseq_cut) 
deseq_res_P2_genus <- runDESeq(P2, deseq_cut)
deseq_res_P3_genus <- runDESeq(P3, deseq_cut)

# print significant taxa
data.table(deseq_res_P1_genus[[2]])
data.table(deseq_res_P2_genus[[2]])
data.table(deseq_res_P3_genus[[2]])
#"ASV_1669" present twice timepoint 1 and 3 

# save
saveRDS(deseq_res_P1_genus, file=paste0(output_data, "DESeq/deseq_res_P1_adjp_genus", deseq_cut, ".rds"))
saveRDS(deseq_res_P2_genus, file=paste0(output_data, "DESeq/deseq_res_P2_adjp_genus", deseq_cut, ".rds"))
saveRDS(deseq_res_P3_genus, file=paste0(output_data, "DESeq/deseq_res_P3_adjp_genus", deseq_cut, ".rds"))


#Including timepoint in the formula
deseq_res_time_genus <- runDESeq(ps_no_norm_filt_genus_conglom, deseq_cut, time = TRUE)
datatable(as.data.frame(ps_no_norm_filt_genus_conglom@tax_table[rownames(deseq_res_time_genus[[2]]), ]))
saveRDS(deseq_res_time_genus, file=paste0(output_data, "DESeq/Deseq_time_interaction_adjp_genus", deseq_cut, ".rds"))
```

#### MetagenomeSeq
```{r}
dir.create(paste0(output_data, 'metagenomseq/'))
###Run ZIG model fitting and prediction
run_metagenom_seq<-function(ps,maxit, mcut, time = F){
  p_metag<-phyloseq_to_metagenomeSeq(ps)
  #filtering at least 4 samples 
  p_metag= cumNorm(p_metag, p=0.75)
  normFactor =normFactors(p_metag)
  normFactor =log2(normFactor/median(normFactor) + 1)
  if(time == F){
    mod = model.matrix(~phenotype + normFactor, data = pData(p_metag))
  }
  if(time == T){
    mod = model.matrix(~phenotype + Within.study.sampling.date +normFactor, data = pData(p_metag))
  }
  
  settings =zigControl(maxit =maxit, verbose =FALSE)
  fit =fitZig(obj = p_metag, mod = mod, useCSSoffset = FALSE, control = settings)
  #Note: changed fit$taxa to fit@taxa in light of error (probably from newer metagenomeseq ver.)
  res_fit_nonfiltered <- MRtable(fit, number = length(fit@taxa))
  res_fit<-res_fit_nonfiltered[res_fit_nonfiltered$adjPvalues<mcut,]
  #finally remove the ones that are not with enough samples
  Min_effec_samp<-calculateEffectiveSamples(fit)
  Min_effec_samp<-Min_effec_samp[ names(Min_effec_samp)  %in% rownames(res_fit)] #####there is a bug here 
  #manually removing the ones with "NA"
  res_fit<-res_fit[grep("NA",rownames(res_fit), inv=T),]
  res_fit$Min_sample<-Min_effec_samp
  res_fit<-res_fit[res_fit$`+samples in group 0` >= Min_effec_samp & res_fit$`+samples in group 1` >= Min_effec_samp,]
  return(list(res_fit_nonfiltered, res_fit))
}
P1<-prune_samples(sample_names(ps_no_norm_filt_genus_conglom)[ps_no_norm_filt_genus_conglom@sam_data$Within.study.sampling.date == "Timepoint 1"], ps_no_norm_filt_genus_conglom)
P2<-prune_samples(sample_names(ps_no_norm_filt_genus_conglom)[ps_no_norm_filt_genus_conglom@sam_data$Within.study.sampling.date == "Timepoint 2"], ps_no_norm_filt_genus_conglom)
P3<-prune_samples(sample_names(ps_no_norm_filt_genus_conglom)[ps_no_norm_filt_genus_conglom@sam_data$Within.study.sampling.date == "Timepoint 3"], ps_no_norm_filt_genus_conglom)

zig_res_P1_genus <- run_metagenom_seq(P1,30, mtgseq_cut) # 30The maximum number of iterations for the expectation-maximization algorithm
zig_res_P2_genus <- run_metagenom_seq(P2,30, mtgseq_cut)
zig_res_P3_genus <- run_metagenom_seq(P3,30, mtgseq_cut)

zig_res_time_genus<- run_metagenom_seq(ps_no_norm_filt_genus_conglom,30, mtgseq_cut, time = T)

# print significant taxa
datatable(zig_res_P1_genus[[2]])
datatable(zig_res_P2_genus[[2]])
datatable(zig_res_P3_genus[[2]])
datatable(zig_res_time_genus[[2]])

zig_res_P1_filtered_genus <- data.frame(cbind(zig_res_P1_genus[[2]], tax_table(P1)[rownames(zig_res_P1_genus[[2]]),]))
zig_res_P1_filtered_genus$enriched <- ifelse(zig_res_P1_filtered_genus$phenotypeN < 0, "Aut", "Control")

zig_res_P2_filtered_genus <- data.frame(cbind(zig_res_P2_genus[[2]], tax_table(P2)[rownames(zig_res_P2_genus[[2]]), ]))
zig_res_P2_filtered_genus$enriched <- ifelse(zig_res_P2_filtered_genus$phenotypeN < 0, "Aut", "Control")

zig_res_P3_filtered_genus <- data.frame(cbind(zig_res_P3_genus[[2]], tax_table(P3)[rownames(zig_res_P3_genus[[2]]), ]))
zig_res_P3_filtered_genus$enriched <- ifelse(zig_res_P3_filtered_genus$phenotypeN < 0, "Aut", "Control")

zig_res_time_filtered_genus <- data.frame(cbind(zig_res_time_genus[[2]], tax_table(ps_no_norm_filt_genus_conglom)[rownames(zig_res_time_genus[[2]]), ]))
zig_res_time_filtered_genus$enriched <- ifelse(zig_res_time_filtered_genus$phenotypeN < 0, "Aut", "Control")


saveRDS(zig_res_P1_genus, file=paste0(output_data, "metagenomseq/zig_res_P1_adjp_genus", mtgseq_cut, ".rds"))
saveRDS(zig_res_P2_genus, file=paste0(output_data, "metagenomseq/zig_res_P2_adjp_genus", mtgseq_cut, ".rds"))
saveRDS(zig_res_P3_genus, file=paste0(output_data, "metagenomseq/zig_res_P3_adjp_genus", mtgseq_cut, ".rds"))
saveRDS(zig_res_time_genus, file=paste0(output_data, "metagenomseq/zig_res_time_adjp_genus", mtgseq_cut, ".rds"))

#do we have any in ESV in common?
#Reduce(intersect, list(rownames(zig_res_P1_filtered),rownames(zig_res_P2_filtered),rownames(zig_res_P3_filtered)))
#rownames(zig_res_P1[[2]])[which(rownames(zig_res_P1[[2]]) %in% c(rownames(zig_res_P2[[2]]), rownames(zig_res_P3[[2]])))]
#rownames(zig_res_P2[[2]])[which(rownames(zig_res_P2[[2]]) %in% rownames(zig_res_P3[[2]]))]

metag_res_genus<-cbind(rownames(zig_res_time_genus[[2]]), as(tax_table(ps_no_norm_filt_genus_conglom)[rownames(zig_res_time_genus[[2]]), ], "matrix"))
metag_res_esv_genus <-metag_res_genus
rownames(metag_res_genus) <- NULL

```

#### Ancom
```{r}

runAncom <- function(ps){
  metada_ps<-sample_data(ps)
  metada_ps<-as.data.frame(metada_ps)
  metada_ps$HostName <- as.character(metada_ps$Host.Name)
  metada_ps$phenotype <- as.character(metada_ps$phenotype)
  
  metada_ps<-tibble(metada_ps$HostName, metada_ps$phenotype, rows = rownames(metada_ps))
  colnames(metada_ps) <- c("HostName", "phenotype", "Biospecimen.Barcode")
  res_table_filt<-otu_table(ps)
  res_table_filt<-as.data.frame(res_table_filt)
  
  prepro<-feature_table_pre_process(feature_table = res_table_filt, meta_data = metada_ps, sample_var = "Biospecimen.Barcode", group_var = "phenotype", out_cut = 0.05, zero_cut = 0.90, lib_cut = 1000, neg_lb = FALSE)
  
  feature_table = prepro$feature_table # Preprocessed feature table
  meta_data = prepro$meta_data # Preprocessed metadata
  struc_zero = prepro$structure_zeros # Structural zero info
  
  main_var = "phenotype"; p_adj_method = "BH"; alpha = 0.05
  adj_formula = NULL; rand_formula = NULL
  ancom_res <- ANCOM(feature_table, meta_data, struc_zero, main_var, p_adj_method, 
                     alpha, adj_formula, rand_formula)#, control)
  return(ancom_res)
}

ancom.all.filt.0.05_genus <- runAncom(ps_no_norm_filt_genus_conglom)
dir.create(path = "ANCOM")
saveRDS(ancom.all.filt.0.05_genus, file=paste0(output_data, "ANCOM/ancom_res_genus_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05_genus$out$detected_0.6)



#Analyze each timepoint independently
#Timepoint 3
ps3<-subset_samples(ps_no_norm_filt_genus_conglom, Within.study.sampling.date == "Timepoint 3" )
ancom.all.filt.0.05.3_genus <- runAncom(ps3)
saveRDS(ancom.all.filt.0.05.3_genus, file=paste0(output_data, "ANCOM/ancom_res3_genus_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.3_genus$out$detected_0.6)
# No sig


#Timepoint 2
ps2<-subset_samples(ps_no_norm_filt_genus_conglom, Within.study.sampling.date == "Timepoint 2" )
ancom.all.filt.0.05.2_genus <- runAncom(ps2)
saveRDS(ancom.all.filt.0.05.2_genus, file=paste0(output_data, "ANCOM/ancom_res2_genus_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.2_genus$out$detected_0.6)


#Timepoint 1
ps1<-subset_samples(ps_no_norm_filt_genus_conglom, Within.study.sampling.date == "Timepoint 1" )
ancom.all.filt.0.05.1_genus <- runAncom(ps1)
saveRDS(ancom.all.filt.0.05.1_genus, file=paste0(output_data, "ANCOM/ancom_res1_genus_", mtgseq_cut, ".rds"))
sum(ancom.all.filt.0.05.1_genus$out$detected_0.6)

```

#### Put together and save results
```{r save_sig_taxa2}
output_data = "../results/"
mtgseq_cut = 0.05
deseq_cut = 0.05
return_dataframe <- function(df){
  tmp = data.frame(asv = rownames(df), padj = df$padj,
                   enrichment = ifelse(df$log2FoldChange < 0, "Aut", "Control"),
                   method = rep("DESeq2", nrow(df)))
  tmp <- tmp %>% mutate_all(as.character)
  return(tmp)
}

return_dataframe_zig <- function(df){
  tmp = data.frame(asv = rownames(df), padj = df$adjPvalues,
                   enrichment = ifelse(df$phenotypeN < 0, "Aut", "Control"),
                   method = rep("ZIG", nrow(df)))
  tmp <- tmp %>% mutate_all(as.character)
  return(tmp)
}

return_dataframe_ancom <- function(df){
  taxa <- as.character(df$taxa_id[df$detected_0.6 == T])
  print(length(taxa))
  if(length(taxa) > 0){
    ps_aut <- prune_samples(ps_css@sam_data$phenotype == 'A', ps_css)
    ps_nt <- prune_samples(ps_css@sam_data$phenotype == 'N', ps_css)
    #find directionality
    t_stat = t.test(as.numeric(ps_aut@otu_table[taxa, ]), as.numeric(ps_nt@otu_table[taxa, ]))$statistic
    enrichment <- ifelse(t_stat > 0, "Aut", "Control") # a t statistic that is positive means group1 is higher than group2. In our case group1 is au
    tmp <- data.frame(asv = taxa, padj = rep(NA, length(taxa)), enrichment = enrichment, method = rep("ANCOM", length(taxa)))
    tmp <- tmp %>% mutate_all(as.character)
    return(tmp)
  }else{
    return(rep(NA, 4))
  }

}
ps_css <- readRDS(paste0(output_data, "Normalized/ps_CSS_pass_min_postDD_min0.03.rds"))

#DESeq results
#deseq_res_P1 <- readRDS(file=paste0(output_data, "DESeq/deseq_res_P1_adjp_genus", deseq_cut, ".rds"))[[2]]
#deseq_res_P1 <- return_dataframe(deseq_res_P1)
#deseq_res_P1$method <- rep("DESEQ_P1", nrow(deseq_res_P1))

#none
#deseq_res_P2 <- readRDS(file=paste0(output_data, "DESeq/deseq_res_P2_adjp_genus", deseq_cut, ".rds"))[[2]]
#deseq_res_P2 <- return_dataframe(deseq_res_P2)
#deseq_res_P2$method <- rep("DESEQ_P2", nrow(deseq_res_P2))

#deseq_res_P3 <- readRDS( file=paste0(output_data, "DESeq/deseq_res_P3_adjp_genus", deseq_cut, ".rds"))[[2]]
#deseq_res_P3 <- return_dataframe(deseq_res_P3)
#deseq_res_P3$method <- rep("DESEQ_P3", nrow(deseq_res_P3))

deseq_res_time <- readRDS( file=paste0(output_data, "DESeq/Deseq_time_interaction_adjp_genus", deseq_cut, ".rds"))[[2]]
deseq_res_time <- return_dataframe(deseq_res_time)
deseq_res_time$method <- rep("DESEQ_time", nrow(deseq_res_time))

deseq_df <- deseq_res_time
deseq_df <- deseq_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))

#Metagenome_seq results
zig_res_P1 <- readRDS( file=paste0(output_data, "metagenomseq/zig_res_P1_adjp_genus", mtgseq_cut, ".rds"))[[2]]
zig_res_P1 <- return_dataframe_zig(zig_res_P1)
zig_res_P1$method <- rep("ZIG_P1", nrow(zig_res_P1))

zig_res_P2 <- readRDS(file=paste0(output_data, "metagenomseq/zig_res_P2_adjp_genus", mtgseq_cut, ".rds"))[[2]]
zig_res_P2 <- return_dataframe_zig(zig_res_P2)
zig_res_P2$method <- rep("ZIG_P2", nrow(zig_res_P2))

zig_res_P3 <- readRDS( file=paste0(output_data, "metagenomseq/zig_res_P3_adjp_genus", mtgseq_cut, ".rds"))[[2]]
zig_res_P3 <- return_dataframe_zig(zig_res_P3)
zig_res_P3$method <- rep("ZIG_P3", nrow(zig_res_P3))

zig_res_time <- readRDS( file=paste0(output_data, "metagenomseq/zig_res_time_adjp_genus", mtgseq_cut, ".rds"))[[2]]
zig_res_time <- return_dataframe_zig(zig_res_time)
zig_res_time$method <- rep("ZIG_time", nrow(zig_res_time))

zig_df <- rbind(zig_res_P1, zig_res_P2, zig_res_P3, zig_res_time)
zig_df <- zig_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))

#Ancom results

#Empty
#ancom_res_P1 <- readRDS(file=paste0(output_data, "ANCOM/ancom_res1_", mtgseq_cut, ".rds"))$out
#ancom_res_P1 <- return_dataframe_ancom(ancom_res_P1)
#ancom_res_P1$method <- rep("ANCOM_P1", nrow(ancom_res_P1))

#Empty
#ancom_res_P2 <- readRDS( file=paste0(output_data, "ANCOM/ancom_res2_", mtgseq_cut, ".rds"))$out
#ancom_res_P2 <- return_dataframe_ancom(ancom_res_P2)
#ancom_res_P2$method <- rep("ANCOM_P2", nrow(ancom_res_P2))

#Empty
#ancom_res_P3 <- readRDS( file=paste0(output_data, "ANCOM/ancom_res3_", mtgseq_cut, ".rds"))$out
#ancom_res_P3 <- return_dataframe_ancom(ancom_res_P3)
#ancom_res_P3$method <- rep("ancom_P3", nrow(ancom_res_P3))

ancom_res_time <- readRDS( file=paste0(output_data, "ANCOM/ancom_res_genus_", mtgseq_cut, ".rds"))$out
ancom_res_time <- return_dataframe_ancom(ancom_res_time)
ancom_res_time$method <- rep("ANCOM_time", nrow(ancom_res_time))

ancom_df <- ancom_res_time #ignore warning and it used to be rbind(ancom_time, p1, etc, but now only time has results)
ancom_df <- ancom_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))
keep <- apply(ancom_df, 1, function(x) return(!is.na(x[1])))
ancom_df <- ancom_df[keep, ]

#all results
results_all <- rbind(deseq_df, zig_df, ancom_df)
results_all <- results_all %>% mutate_all(as.character)
results_grouped <- results_all %>% group_by(asv) %>% summarize(enrichment = paste0(enrichment, collapse = ", "), method = paste0(method, collapse = ", "))
results_grouped <- data.frame(cbind(results_grouped, ps_no_norm_filt_genus_conglom@tax_table[results_grouped$asv, ]))
write.table(results_grouped, paste0(output_data, "significant_taxa_all_genus.tsv"), row.names = F, sep = "\t")
data.table(results_grouped)

#time series
results_df <- rbind(deseq_res_time, zig_res_time, ancom_res_time)
results_df <- results_df %>% mutate_all(as.character)
results_df_grouped <- results_df %>% group_by(asv) %>% summarize(enrichment = list(enrichment), method = paste0(method, collapse = ", "))
#add taxa information
results_df_grouped <- data.frame(cbind(results_df_grouped, ps_no_norm_filt_genus_conglom@tax_table[results_df_grouped$asv, ]))
keep <- apply(results_df_grouped, 1, function(x) return(all(x['enrichment'][[1]] == "Aut") | all(x['enrichment'][[1]] == "Control"))) 
results_final<- results_df_grouped[keep, ]
results_final$enrichment <- unlist(lapply(results_final$enrichment, function(x) return(paste0(x, collapse = ","))))
data.table(results_final)
write.table(results_final, paste0(output_data, "significant_taxa_timeseries_genus.tsv"), row.names = F, sep = "\t")

data.table(results_final[,c(2, 3, 8, 9)])


```

```{r}
sessionInfo()

```
