---
title: "confounders"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(vegan)
library(phyloseq)
library(gridExtra)
library(adegraphics)
library(dplyr)
library(data.table)
library(DT)
library(ggplot2)
library(metagenomeSeq)
library(cowplot)
library(DESeq2)
```

### Load Data
```{r message=FALSE, warning=FALSE}
output_data <- "../results/"
ps_deseq <- readRDS(paste0(output_data, "Filtered/ps_deseq_friedfilt.rds"))
ps_deseq@sam_data$High_Anxiety<-gsub("1", "No" ,ps_deseq@sam_data$Recent.anxiety..caretaker.reported.)
ps_deseq@sam_data$High_Anxiety<-gsub("2", "No", ps_deseq@sam_data$High_Anxiety)
ps_deseq@sam_data$High_Anxiety<-as.factor(gsub("3", "Yes", ps_deseq@sam_data$High_Anxiety))

map<-as.data.frame(sample_data(ps_deseq))

num_cat <- colnames(map)[unlist(lapply(map, is.numeric))]
num_cat <- num_cat[-which(num_cat %in% c("Number.of.small.pet.herbivores", "Number.of.small.pet.rodents", "phenotype_num", "Number.of.pet.birds", "Host.ID", "Biospecimen.ID", "NCBI.taxID" ))]

logical_cat <- colnames(map)[unlist(lapply(map, is.logical))]

for (i in logical_cat){
  map[,which(colnames(map) %in% c(i))]<- as.factor(getVariable(map, i))
}

fac_cat <- colnames(map)[unlist(lapply(map, is.factor))]
#finally removing the ones that were only asked for the children with ASD, or only have one factor & NA, or only present in one phen
fac_cat<-fac_cat[-which(fac_cat %in% c("Behavior.video.submitted..M3.","Language.ability.and.use","Conversation.ability","Understands.speech","Plays.imaginatively.when.alone","Plays.imaginatively.with.others","Plays.in.a.group.with.others","Eye.contact.finding","Childhood.behavioral.development.finding","Repetitive.motion","Picks.up.objects.to.show.to.others","Sleep.pattern.finding","Response.to.typical.sounds","Self.injurious.behavior.finding","Gastrointestinal.problems..M3.", "Imitation.behavior", "Other.stool.sample.collection.method.explained..M3.", "Flu.shot.in.the.last..MFlu.shot.in.the.last..M3.", "Pica.disease", "Additional.info.affecting.microbiome..M3.", "Dietary.restrictions.details..M3.", "Pet.bird", "Host.interaction.with.pets" ))]



```

# PERMANOVA
non- parametric statistical approaches (ANOSIM, ADONIS, ANOVA, PERMANOVA, etc.) will be employed to determine the significance of noteworthy factors, such as digital phenotype, probiotic and/or antibiotic use

## PERMANOVA test
```{r permanova, results="asis"}
permanova <- function(physeq, factorName, ifnumeric, pmt = 999){
  set.seed(1)
  braydist = phyloseq::distance(physeq, "bray")
  form <- as.formula(paste("braydist ~ ", c(factorName), sep = ""))
  metaDF=data.frame(sample_data(physeq)[, as.character(factorName)])
  # if numerical variable, make sure the class
  if(ifnumeric){
    metaDF[, factorName] <- as.numeric(metaDF[, factorName])
    factor_class='numeric'
  } else {
    factor_class='categorical'
  }
  perm <- adonis(form, permutations = pmt, metaDF)
  permDT=data.table(Variable=factorName, 
             FactorClass=factor_class,
             TotalN=perm$aov.tab['Total','Df']+1, 
             R2=perm$aov.tab[factorName, 'R2'], 
             pvalue=perm$aov.tab[factorName,'Pr(>F)'][1])
  return(permDT)
}

#we keep only the cateory selected above as relevant 
tmp_metadata <- map[,c(num_cat,fac_cat)]
#For categorical variables, keep only those with 2 or more possible answers
cat_drop <- sapply(tmp_metadata[ , fac_cat], function(x) return(length(levels(x)) < 2))
fac_cat_drop <- fac_cat[cat_drop]
tmp_metadata <- tmp_metadata[, -which(colnames(tmp_metadata) %in% fac_cat_drop)]
#For all variables, keep only those where at least 20% have valid answers
cols_keep <- as.vector(apply(tmp_metadata, 2, function(x) return(sum(is.na(x)) < length(x)*.8 )))
metadata <- tmp_metadata[ , cols_keep]

# filter out variables with n-4 all same responses across datset (140 out of 144 samples (at least 420 same responses))
cols_keep <- as.vector(apply(metadata, 2, function(x) return(sum(table(x)[which(table(x) >= 420)] >= 420)))) == 0
metadata <- metadata[ , cols_keep]


#Test the homogeneity of dispersion before running permanova - we only wish to test variables that are dispersed homogeneously with respect
#to the distances between samples in microbiome space
set.seed(1)
pval_factors_disper=c()
nb_samples_disper=c()
for (i in 1:ncol(metadata)){
  cat (i,"\t")
  #Drop samples that have NA for the specified variable
  test_map <- metadata[!is.na(metadata[,i]) & metadata[,i] != "" ,]
  ps_tmp <- ps_deseq
  sample_data(ps_tmp) <- test_map

  dist_bray <- phyloseq::distance(ps_tmp, method = "bray")
  beta <- betadisper(dist_bray, get_variable(ps_tmp)[, i])
  tmp <- permutest(beta)
  tmp <- tmp$tab$`Pr(>F)`[1]
  pval_factors_disper <- c(pval_factors_disper,tmp)
  nb_samples_disper <- c(nb_samples_disper, nsamples(ps_tmp))}

#correct the p.value 
names(pval_factors_disper) <- colnames(metadata)
pval_factors_disper <- p.adjust(pval_factors_disper, method = "fdr")
to_remove_betadis <- names(pval_factors_disper)[pval_factors_disper < 0.05]

permanova_df <- metadata[,-which(colnames(metadata) %in% to_remove_betadis)]
set.seed(1)
permanova_res=NULL
for(var_name in colnames(permanova_df)){
  print(var_name)
  #Drop samples that have NA for the specified variable
  test_map <- permanova_df[!is.na(permanova_df[ , var_name]) & permanova_df[ , var_name] != "" , ]
  ps_tmp <- ps_deseq
  sample_data(ps_tmp) <- test_map
  
  # Check if there is more than 1 values (categories)
  if(uniqueN(ps_tmp@sam_data[ , var_name]) > 1){
    
    # if categorical
    if(is.factor(permanova_df[[var_name]])){
      # run permanova only if there are more than 1 groups
      p <- permanova(ps_tmp, factorName = var_name, ifnumeric = FALSE, pmt = 999)
      permanova_res=rbindlist(list(permanova_res, p))
      rm(p)
    }
    # if continuous
    if(is.numeric(permanova_df[[var_name]])){
      p <- permanova(ps_tmp, factorName = var_name, ifnumeric = TRUE, pmt = 999)
      permanova_res=rbindlist(list(permanova_res, p))
      rm(p)
    }
  }
  rm(var_name)
}

#Correct p_value with fdr (added Oct_14_2020)
permanova_res$pvalue_adj<-  p.adjust(permanova_res$pvalue, method = "fdr")

#removing LR predictions since those are essentially an indicator of phenotype and not confounding variables
permanova_res <- permanova_res[!(permanova_res$Variable %in% c("LR10.probability.ASD..M3.", "LR5.probability.ASD..M3.", "LR6.probability.ASD..M3." , "LR10.prediction..M3.", "LR10.probability.not.ASD..M3.", "LR5.probability.not.ASD..M3." , "LR5.prediction..M3.", "LR6.prediction..M3." , "LR6.probability.not.ASD..M3.")), ]

#permanova_res <- permanova_res[!(permanova_res$Variable %in% c("Language.ability.and.use", "Conversation.ability", #"Plays.imaginatively.when.alone", "Plays.imaginatively.with.others", "Plays.in.a.group.with.others", #"Eye.contact.finding", "Picks.up.objects.to.show.to.others", "Imitation.behavior","Understands.speech", #"Childhood.behavioral.development.finding","Repetitive.motion", "Sleep.pattern.finding", "Response.to.typical.sounds", #"Self.injurious.behavior.finding", "Gastrointestinal.problems..M3.")),]

# sort
permanova_res <- permanova_res[order(permanova_res$R2, decreasing=TRUE), ]
permanova_res$R2 <- round(permanova_res$R2, digits = 3)
permanova_res$pvalue_adj <- round(permanova_res$pvalue_adj, digits = 3)
datatable(permanova_res)
write.csv(permanova_res, file=paste0(output_data, 'PERMANOVA_noLR.csv'), row.names=FALSE)

```



## Visualize permanova significant variables on PCoA
```{r plot_R2_value, fig.width=8.5, fig.height=11}
# function to plot PCoA, only for higher R2 value 
permanova_res <- permanova_res[permanova_res$pvalue_adj < 0.05,]

imp_factors<-permanova_res$Variable#[permanova_res$R2 > 0.009]
imp_factors <- gsub(".1","", imp_factors)

#Drop factors that have too many NA values
keep_factors <- imp_factors[apply(map[, imp_factors], 2 ,function(x) return(sum(is.na(x)) < 30))]
#Drop samples that have NA for any of the factors of interest
map_plot <- map[ , keep_factors]
map_plot <- map_plot[complete.cases(map_plot), ]

ps_deseq_pcoa <- ps_deseq
sample_data(ps_deseq_pcoa) <- map_plot
f <- as.formula(paste0("~", paste0(colnames(map_plot), collapse = "+")))

#ordination formula .
ord_pcoa <- ordinate(
  physeq = ps_deseq_pcoa, 
  method = "CAP", 
  distance = "bray",
  formula = f)

title_prep <- c(colnames(map_plot), "phenotype", "Annual.household.income_rank")
to_plot=list()
for (i in 1:length(title_prep)){
  to_plot[[i]] <- plot_ordination(
  physeq = ps_deseq, 
  ordination = ord_pcoa, 
  color = title_prep[i], 
  axes = c(1,2),
  title=title_prep[i]
) + 
 geom_point( size = 0.15) +
 theme(text = element_text(size =7.5), plot.title = element_text(size=10), legend.title = element_blank())
}
to_plot[[length(title_prep)+1]] <- plot_ordination(physeq = ps_deseq, ordination = ord_pcoa, type="taxa",title ="Taxa") + theme(text = element_text(size =5), legend.title = element_blank())

pdf(paste0(output_data,"confounding_factors.pdf"))
#grid.arrange(grobs = as.list(to_plot), layout_matrix = lay)
plot_grid(plotlist = to_plot[c(1:4)], ncol = 2)
plot_grid(plotlist = to_plot[c(5:8)], ncol = 2)
plot_grid(plotlist = to_plot[c(9:12)], ncol = 2)
plot_grid(plotlist = to_plot[c(13:16)], ncol = 2)
plot_grid(plotlist = to_plot[c(17:20)], ncol = 2)
plot_grid(plotlist = to_plot[c(21:24)], ncol = 2)
plot_grid(plotlist = to_plot[c(25:28)], ncol = 2)
plot_grid(plotlist = to_plot[c(29:31)], ncol = 2)
dev.off()
top_potential_confounds <- keep_factors
top_potential_confounds

```


## With these variables added into ordination
```{r}
i = "phenotype"
f <- as.formula(paste0("~", paste0(colnames(map_plot), collapse = "+"),"+", i))

#Drop samples that have NA for any of the factors of interest
map_plot <- map[ , c(keep_factors, "phenotype", "Annual.household.income_rank")]
map_plot <- map_plot[complete.cases(map_plot), ]

ps_deseq_pcoa <- ps_deseq
sample_data(ps_deseq_pcoa) <- map_plot

#ordination formula
ps_deseq_pcoa
ord_pcoa <- ordinate(
  physeq = ps_deseq_pcoa, 
  method = "CAP", 
  distance = "bray",
  formula = f)
plot_ordination(
  physeq = ps_deseq, 
  ordination = ord_pcoa, 
  color = i, 
  axes = c(1,2),
  title=i
) + 
  geom_point( size = 0.5) +
  theme(text = element_text(size =20), plot.title = element_text(size=15))

i = "Annual.household.income_rank"
f <- as.formula(paste0("~", paste0(colnames(map_plot), collapse = "+"), "+", i))
ord_pcoa <- ordinate(
  physeq = ps_deseq_pcoa, 
  method = "CAP", 
  distance = "bray",
  formula = f)
plot_ordination(
  physeq = ps_deseq, 
  ordination = ord_pcoa, 
  color = i, 
  axes = c(1,2),
  title=i
) + 
  geom_point( size = 0.5) +
  theme(text = element_text(size =20), plot.title = element_text(size=15), legend.title = element_blank())

```

```{r plot_imp_taxa}
#Let's have a look at the plot 
plot_ordination(physeq = ps_deseq, ordination = ord_pcoa, type="taxa",title ="Taxa") + theme(text = element_text(size =8))
#ok let's try to find the spcies that show some importance in this PCA
taxa.to.select<-vegan::scores(ord_pcoa)$species
#now plot it with no name for visibilty
rownames(taxa.to.select)<-c()
s.arrow(taxa.to.select) #the taxa that influence the most the plots are above 0.1
scores = vegan::scores(ord_pcoa)$species
taxa.to.select.to.rem <- scores[as.vector(abs(scores[,1])>0.1 | abs(scores[,2])>0.1),]

#any overlap with the 5 important? 
deseq_res <- readRDS(paste0(output_data, "DESeq/full_res_table_deseq.rds"))
rownames(deseq_res) %in% taxa.to.select.to.rem #NOPE!

saveRDS(taxa.to.select.to.rem, paste0(output_data, "taxa_associated_confounding_variables.rds"))
```

```{r}
run_metagenom_seq<-function(ps, variable, maxit, mcut, time = F){
  ps <- prune_samples(as.vector(!is.na(ps@sam_data[ , variable])), ps)
  p_metag<-phyloseq_to_metagenomeSeq(ps)
  #filtering at least 4 samples 
  p_metag= cumNorm(p_metag, p=0.75)
  normFactor =normFactors(p_metag)
  normFactor =log2(normFactor/median(normFactor) + 1)
  if(time == F){
    form <- as.formula(paste("~", variable, "+ normFactor" ))
    mod = model.matrix(form, data = pData(p_metag))
  }
  if(time == T){
    form <- as.formula(paste("~", variable, " + Within.study.sampling.date + normFactor" ))
    mod = model.matrix(form, data = pData(p_metag))
  }
  
  settings =zigControl(maxit =maxit, verbose =FALSE)
  fit =fitZig(obj = p_metag, mod = mod, useCSSoffset = FALSE, control = settings)
  #Note: changed fit$taxa to fit@taxa in light of error (probably from newer metagenomeseq ver.)
  res_fit_nonfiltered <- MRtable(fit, number = length(fit@taxa))
  res_fit<-res_fit_nonfiltered[res_fit_nonfiltered$adjPvalues<mcut,]
  #finally remove the ones that are not with enough samples
  Min_effec_samp<-calculateEffectiveSamples(fit)
  Min_effec_samp<-Min_effec_samp[ names(Min_effec_samp)  %in% rownames(res_fit)] #####there is a bug here 
  #manually removing the ones with "NA"
  res_fit<-res_fit[grep("NA",rownames(res_fit), inv=T),]
  res_fit$Min_sample<-Min_effec_samp
  res_fit<-res_fit[res_fit$`+samples in group 0` >= Min_effec_samp & res_fit$`+samples in group 1` >= Min_effec_samp,]
  return(list(res_fit_nonfiltered, res_fit))
}
```
## MARA ordinated variables only in ASD cohort

```{r}

#we keep only the cateory selected above as relevant 
tmp_metadata <- map[,c(num_cat,fac_cat)]

#scores for social, learning, and anxiety
social <- c("Language.ability.and.use", "Conversation.ability", "Plays.imaginatively.when.alone", "Plays.imaginatively.with.others", "Plays.in.a.group.with.others", "Eye.contact.finding", "Picks.up.objects.to.show.to.others", "Imitation.behavior")

learning <- c("Understands.speech", "Childhood.behavioral.development.finding")

anxiety <- c("Repetitive.motion", "Sleep.pattern.finding", "Response.to.typical.sounds", "Self.injurious.behavior.finding", "Recent.anxiety..caretaker.reported.")

#only ASD
tmp_metadata<-tmp_metadata[tmp_metadata$Host.disease.status == "Case (disease status)",]
ps_deseq_ASD <-subset_samples(ps_deseq, phenotype == "A")

#MARa variables
cols_keep <- c(anxiety[-which(anxiety %in% "Recent.anxiety..caretaker.reported.")], learning, social)
metadata <- tmp_metadata[ , c(cols_keep, "Mobile.Autism.Risk.Assessment.Score")]



#Test the homogeneity of dispersion before running permanova - we only wish to test variables that are dispersed homogeneously with respect
#to the distances between samples in microbiome space
set.seed(1)
pval_factors_disper=c()
nb_samples_disper=c()
for (i in 1:ncol(metadata)){
  cat (i,"\t")
  #Drop samples that have NA for the specified variable
  test_map <- metadata[!is.na(metadata[,i]) & metadata[,i] != "" ,]
  ps_tmp <- ps_deseq_ASD
  sample_data(ps_tmp) <- test_map

  dist_bray <- phyloseq::distance(ps_tmp, method = "bray")
  beta <- betadisper(dist_bray, get_variable(ps_tmp)[, i])
  tmp <- permutest(beta)
  tmp <- tmp$tab$`Pr(>F)`[1]
  pval_factors_disper <- c(pval_factors_disper,tmp)
  nb_samples_disper <- c(nb_samples_disper, nsamples(ps_tmp))}

#correct the p.value 
names(pval_factors_disper) <- colnames(metadata)
pval_factors_disper <- p.adjust(pval_factors_disper, method = "fdr")
to_remove_betadis <- names(pval_factors_disper)[pval_factors_disper < 0.05]

permanova_df <- metadata[,-which(colnames(metadata) %in% to_remove_betadis)]
set.seed(1)
permanova_res=NULL
for(var_name in colnames(permanova_df)){
  print(var_name)
  #Drop samples that have NA for the specified variable
  test_map <- permanova_df[!is.na(permanova_df[ , var_name]) & permanova_df[ , var_name] != "" , ]
  ps_tmp <- ps_deseq_ASD
  sample_data(ps_tmp) <- test_map
  
  # Check if there is more than 1 values (categories)
  if(uniqueN(ps_tmp@sam_data[ , var_name]) > 1){
    
    # if categorical
    if(is.factor(permanova_df[[var_name]])){
      # run permanova only if there are more than 1 groups
      p <- permanova(ps_tmp, factorName = var_name, ifnumeric = FALSE, pmt = 999)
      permanova_res=rbindlist(list(permanova_res, p))
      rm(p)
    }
    # if continuous
    if(is.numeric(permanova_df[[var_name]])){
      p <- permanova(ps_tmp, factorName = var_name, ifnumeric = TRUE, pmt = 999)
      permanova_res=rbindlist(list(permanova_res, p))
      rm(p)
    }
  }
  rm(var_name)
}

#Correct p_value with fdr (added Oct_14_2020)
permanova_res$pvalue_adj<-  p.adjust(permanova_res$pvalue, method = "fdr")


# sort
permanova_res <- permanova_res[order(permanova_res$R2, decreasing=TRUE), ]
permanova_res$R2 <- round(permanova_res$R2, digits = 3)
permanova_res$pvalue_adj <- round(permanova_res$pvalue_adj, digits = 3)
datatable(permanova_res[permanova_res$pvalue_adj <= 0.05])
write.csv(permanova_res, file ="../results/MARA_var_perma.csv") 

#curiosity
mara_adonis<-permanova(ps_deseq_ASD, factorName = "Mobile.Autism.Risk.Assessment.Score", ifnumeric = TRUE, pmt = 999)

#common
total <- read.csv('../results/PERMANOVA_noLR.csv')
permanova_res$Variable %in% total$Variable
```
## For each significant variable in permanova test, use DESeq and metagenomeseq to identify the taxa relevant to that variable
```{r}
permanova_res <- read.csv('../results/PERMANOVA_noLR.csv')
permanova_res <- permanova_res[permanova_res$pvalue_adj < 0.05,]
MARA <-read.csv(file ="../results/MARA_var_perma.csv")
behavior <-c(anxiety[-which(anxiety %in% "Recent.anxiety..caretaker.reported.")], learning, social)
take_out<-behavior#[-which(behavior %in% MARA$Variable)] 
if(nrow(permanova_res[-which(permanova_res$Variable %in% take_out),]) != 0){
permanova_res<-permanova_res[-which(permanova_res$Variable %in% take_out),]
}
write.csv(permanova_res,"../filtered_perm_res.csv")

ps_no_norm_filt <- readRDS(paste0(output_data, "Filtered/ps_no_norm_friedfilt.rds"))
taxa_associated_w_variables <- list()
perma_var<-permanova_res$Variable
for(var in perma_var[1:length(perma_var)]){
  print(var)
  tmp <- run_metagenom_seq(ps_no_norm_filt, variable = as.character(var), 30, .05, time = T)
  if (is.numeric(get_variable(ps_no_norm_filt, var)) == FALSE){
  taxa_associated_w_variables[[var]] <- data.frame(rownames(tmp[[2]]), tmp[[2]]$adjPvalues, rep(var, nrow(tmp[[2]])), tmp[[2]][, paste0(var, "TRUE")] >= 0)
  colnames(taxa_associated_w_variables[[var]]) <- c("ASV", "adjPvalue", "variable", "enrichment")} else{
  taxa_associated_w_variables[[var]] <- data.frame(rownames(tmp[[2]]), tmp[[2]]$adjPvalues, rep(var, nrow(tmp[[2]])), tmp[[2]][, var] >= 0)
  colnames(taxa_associated_w_variables[[var]]) <- c("ASV", "adjPvalue", "variable", "enrichment")
  }
}
keep <- as.numeric(sapply(taxa_associated_w_variables, nrow)) > 0
taxa_associated_w_variables_keep <- do.call(rbind, taxa_associated_w_variables[keep])
#colnames(taxa_associated_w_variables_keep) <- c("ASV", "adjPvalue", "variable", "enrichment")

write.csv(taxa_associated_w_variables_keep, paste0(output_data, "taxa_associated_w_variables.csv"))

taxa_associated_w_high_anx <- list()
perma_var<-"High_Anxiety"
ps_no_norm_filt@sam_data$High_Anxiety<-gsub("1", "No" ,ps_no_norm_filt@sam_data$Recent.anxiety..caretaker.reported.)
ps_no_norm_filt@sam_data$High_Anxiety<-gsub("2", "No", ps_no_norm_filt@sam_data$High_Anxiety)
ps_no_norm_filt@sam_data$High_Anxiety<-as.factor(gsub("3", "Yes", ps_no_norm_filt@sam_data$High_Anxiety))
for(var in perma_var[1:length(perma_var)]){
  print(var)
  tmp <- run_metagenom_seq(ps_no_norm_filt, variable = as.character(var), 30, .05, time = T)
  taxa_associated_w_high_anx[[var]] <- data.frame(rownames(tmp[[2]]), tmp[[2]]$adjPvalues, rep(var, nrow(tmp[[2]])))
}
#keepanx <- as.numeric(sapply(taxa_associated_w_high_anx, nrow)) > 0
#taxa_associated_w_high_anx_keep <- do.call(rbind, taxa_associated_w_high_anx[keepanx])
#colnames(taxa_associated_w_high_anx_keep) <- c("ASV", "adjPvalue", "variable")
#see later lines for comparison to 117
```

#Add to significant taxa chart
```{r}
addConfounders <- function(sig_taxa, taxa_ass_w_var){
  sig_taxa$associated_variables <- rep(NA, nrow(sig_taxa))
  for(asv in rownames(sig_taxa)){
    associated_variables <- paste0(rownames(taxa_ass_w_var[taxa_ass_w_var$ASV ==asv, ]), taxa_ass_w_var[taxa_ass_w_var$ASV ==asv, ]$enrichment)
    associated_variables <- gsub('[[:digit:]]+', '', associated_variables)
    sig_taxa[asv, ]$associated_variables <- paste(associated_variables, collapse= ", ")
  }
  sig_taxa$associated_variables <- gsub("Host.disease.status", "", sig_taxa$associated_variables)
  return(sig_taxa)
}
```

```{r}
taxa_ass_w_var <- read.delim(paste0(output_data, "taxa_associated_w_variables.csv"), sep = ",", row.names = 1)
taxa_ass_w_var$ASV <- as.character(taxa_ass_w_var$ASV)

sig_taxa <- read.delim(paste0(output_data, "significant_taxa_timeseries.tsv"), sep = '\t', row.names = 1)
sig_taxa <- addConfounders(sig_taxa, taxa_ass_w_var)
write.csv(sig_taxa, paste0(output_data, "significant_taxa_timeseries_confounders.csv"))

sig_taxa <- read.delim(paste0(output_data, "significant_taxa_all.tsv"), sep = '\t', row.names = 1)
sig_taxa <- addConfounders(sig_taxa, taxa_ass_w_var)
write.csv(sig_taxa, paste0(output_data, "significant_taxa_all_confounders.csv"))

```

## Breastfed check
```{r}
ps_all<-readRDS("../results/Filtered/ps_allsamps_deseq_norm_friedfilt.rds")

breast_fed <- c("089_A","054_N", "158_N" )
ps_all@sam_data$breastfed <- rep("FALSE", length(rownames(ps_all@sam_data)))
ps_all@sam_data$breastfed[which(ps_all@sam_data$Host.Name %in% breast_fed)] <- rep("TRUE", length(ps_all@sam_data$breastfed[which(ps_all@sam_data$Host.Name %in% breast_fed)]))

tab<-permanova(ps_all, "breastfed", "FALSE")

dist_bray <- phyloseq::distance(ps_all, method = "bray")
beta <- betadisper(dist_bray, get_variable(ps_all)[, "breastfed"])
tmp <- permutest(beta)
tmp <- tmp$tab$`Pr(>F)`[1]

tab$betadispr_pval <- tmp

```


## MARA ordinated variables only in ASD cohort plot

```{r,  fig.width=16, fig.height=40}

#we keep only the cateory selected above as relevant 
tmp_metadata <- map[,c(num_cat,fac_cat)]

#scores for social, learning, and anxiety
social <- c("Language.ability.and.use", "Conversation.ability", "Plays.imaginatively.when.alone", "Plays.imaginatively.with.others", "Plays.in.a.group.with.others", "Eye.contact.finding", "Picks.up.objects.to.show.to.others", "Imitation.behavior")

learning <- c("Understands.speech", "Childhood.behavioral.development.finding")

anxiety <- c("Repetitive.motion", "Sleep.pattern.finding", "Response.to.typical.sounds", "Self.injurious.behavior.finding", "Recent.anxiety..caretaker.reported.")

#only ASD
tmp_metadata<-tmp_metadata[tmp_metadata$Host.disease.status == "Case (disease status)",]
ps_deseq_ASD <-subset_samples(ps_deseq, phenotype == "A")

#MARa variables
cols_keep <- c(anxiety[-which(anxiety %in% "Recent.anxiety..caretaker.reported.")], learning, social)
take_out<-behavior[-which(behavior %in% MARA$Variable)] 
cols_keep<-cols_keep[-which(cols_keep %in% take_out)]
metadata <- tmp_metadata[ , cols_keep]

metadata <- metadata[complete.cases(metadata), ]

ps_deseq_pcoa <- ps_deseq_ASD
sample_data(ps_deseq_pcoa) <- metadata
f <- as.formula(paste0("~", paste0(colnames(metadata), collapse = "+")))

#ordination formula .
ord_pcoa <- ordinate(
  physeq = ps_deseq_pcoa, 
  method = "CAP", 
  distance = "bray",
  formula = f)

title_prep <- colnames(metadata)
to_plot=list()
for (i in 1:length(title_prep)){
  to_plot[[i]] <- plot_ordination(
  physeq = ps_deseq_pcoa, 
  ordination = ord_pcoa, 
  color = title_prep[i], 
  axes = c(1,2),
  title=title_prep[i]
) + 
  geom_point( size = 0.5) +
  theme(text = element_text(size =20), plot.title = element_text(size=15))
}

lay <- rbind(c(1),
             c(2),
             c(3),
             c(4),
             c(5),
             c(6),
             c(7),
             c(8),
             c(9), 
             c(10))


#pdf(paste0(output_data,"confounding_factors.pdf",width=16,height=40))
grid.arrange(grobs = as.list(to_plot), layout_matrix = lay)


```

## Find associated taxa for each significant behavioral trait
```{r}
permanova_res<-read.csv(file ="../results/MARA_var_perma.csv", row.names = 1) 
perma_var<-permanova_res$Variable
ps_no_norm_filt <- readRDS(paste0(output_data, "Filtered/ps_no_norm_friedfilt.rds"))

ps_no_norm_filt@sam_data$High_Anxiety<-gsub("1", "FALSE" ,ps_no_norm_filt@sam_data$Recent.anxiety..caretaker.reported.)
ps_no_norm_filt@sam_data$High_Anxiety<-gsub("2", "FALSE", ps_no_norm_filt@sam_data$High_Anxiety)
ps_no_norm_filt@sam_data$High_Anxiety<-gsub("3", "TRUE", ps_no_norm_filt@sam_data$High_Anxiety)
ps_no_norm_filt@sam_data$High_Anxiety <- as.factor(ps_no_norm_filt@sam_data$High_Anxiety)
#perma_var <- c(perma_var, "High_Anxiety")
ps_ASD<-subset_samples(ps_no_norm_filt, phenotype == "A")

taxa_associated_w_variables <- list()
for(var in perma_var[1:length(perma_var)]){
  print(var)
  tmp <- run_metagenom_seq(ps_ASD, variable = var, 30, .05, time = T)
  taxa_associated_w_variables[[var]] <- data.frame(rownames(tmp[[2]]), tmp[[2]]$adjPvalues, rep(var, nrow(tmp[[2]])))
}
keep <- as.numeric(sapply(taxa_associated_w_variables, nrow)) > 0
taxa_associated_w_variables_keep <- do.call(rbind, taxa_associated_w_variables[keep])
if (!is.null(taxa_associated_w_variables_keep)){
colnames(taxa_associated_w_variables_keep) <- c("ASV", "adjPvalue", "variable")
}
write.csv(taxa_associated_w_variables_keep, paste0(output_data, "taxa_associated_w_behavioral_variables.csv"))
taxa_associated_w_variables_keep

sig_res <- read.table(paste0(output_data, "significant_taxa_timeseries.tsv"), row.names = 1, sep = "\t", header = T)

sig_res[which(rownames(sig_res) %in% taxa_associated_w_variables_keep$ASV),]
high_anx_metagenomeseq<-sig_res[which(rownames(sig_res) %in% taxa_associated_w_variables_keep$ASV),]
high_anx_metagenomeseq
#write.csv(high_anx_metagenomeseq, "../High_anx_mtg_in_common_w_117.csv")

#whole datset
#sig_res[which(rownames(sig_res) %in% taxa_associated_w_high_anx_keep$ASV),]
```

#DESEq model using significant permanova and chi-square significant

```{r}
ps_not_norm_filt <- readRDS(paste0(output_data, "Filtered/ps_no_norm_friedfilt.rds"))

ps_no_norm_filt@sam_data$High_Anxiety<-gsub("1", "0" ,ps_no_norm_filt@sam_data$Recent.anxiety..caretaker.reported.)
ps_no_norm_filt@sam_data$High_Anxiety<-gsub("2", "0", ps_no_norm_filt@sam_data$High_Anxiety)
ps_no_norm_filt@sam_data$High_Anxiety<-gsub("3", "1", ps_no_norm_filt@sam_data$High_Anxiety)

#3 gone here
ps_allperma<-subset_samples(ps_no_norm_filt, !is.na(Seafood..consumption.frequency.))
ps_allperma<-subset_samples(ps_allperma, !is.na(Dairy..consumption.frequency...longitudinal.))
ps_allperma<-subset_samples(ps_allperma, !is.na(Dietary.supplement))



runDESeq <- function(ps, dcut, time = FALSE){
  if(time == TRUE){
    #Working with time series
    #according to the DeSeq vignette: design including the time factor, and then test using the likelihood ratio test as described
    #the following section, where the time factor is removed in the reduced formula
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype + Within.study.sampling.date + Seafood..consumption.frequency. + Dairy..consumption.frequency...longitudinal. + Vegetable..consumption.frequency...longitudinal. + Dietary.supplement + GI.symptoms.within.3.months..M3. + GI.issues.this.week..M3. + High_Anxiety) 
  }else{
    diagdds = phyloseq_to_deseq2(ps, ~ phenotype) 
  }
  
  #diagdds <- estimateSizeFactors(diagdds, type = "poscounts")
  diagdds <- DESeq(diagdds,fitType="parametric", betaPrior = FALSE) 
  res = results(diagdds, contrast = c("phenotype", "N", "A"))
  res$padj[is.na(res$padj)] = 1
  sig <- res[res$padj < dcut,]
  if (dim(sig)[1] == 0) 
  {sigtab<- as.data.frame(1, row.names="nothing")
  colnames(sigtab) <- 'padj'}
  else 
  {
    sigtab <- data.frame(sig)
  }
  return(list(res, sigtab))
}

deseq_res_Perma<-runDESeq(ps_allperma, dcut = 0.05, time = TRUE)

#87 hits
data.table(deseq_res_Perma[[2]])

sig_res <- read.table(paste0(output_data, "significant_taxa_timeseries.tsv"), row.names = 1, sep = "\t", header = T)
#8 in common with 117
sig_res[which(rownames(sig_res) %in% rownames(deseq_res_Perma[[2]])),]

use <- grepl(",", sig_res$method)
eleven_two_methods<-sig_res[use, ]

#Only g__Bacteroides	s__thetaiotaomicron
eleven_two_methods[which(rownames(eleven_two_methods) %in% rownames(deseq_res_Perma[[2]])),]


#find amount of sig asvs associated with a lifestyle variable associated with ASD or TD
a<-read.csv("../results/significant_taxa_timeseries_confounders.csv")
b <- read.csv("../results/taxa_associated_w_variables.csv")
sum(a$X %in% b$ASV)

permav_res<-read.csv("../filtered_perm_res.csv")
sigvars<-permav_res$Variable

#fix Gi symptoms one due to gsub mistaken change
sigvars<-gsub("GI.symptoms.within.3.months..M3.","GI.symptoms.within..months", sigvars )

#sigvariables_in_perma_and_chi <- c("Dietary.restrictions", "Dietary.supplement", "Dairy..consumption.frequency...longitudinal", "Vegetable..consumption.frequency...longitudinal", "GI.symptoms.within..months", "GI.issues.this.week")

rowlist <- list()
for (i in 1:length(sigvars)) {
  rowlist[[i]]<-grep(sigvars[i], a$associated_variables)
  
}

length(unique(unlist(rowlist)))

unique(unlist(rowlist))

#find which of the 72 are enriched in ASD and TD
a$X[unique(unlist(rowlist))]

a$enrichment[unique(unlist(rowlist))]

length(grep("Aut", a$enrichment[unique(unlist(rowlist))]))

length(grep("Control", a$enrichment[unique(unlist(rowlist))]))

unique(b$ASV)[-which(unique(b$ASV) %in% a$X[unique(unlist(rowlist))])]

#find which of the remaining unassociated 45 are enriched in ASD and TD
a$enrichment[-which(a$X %in% a$X[unique(unlist(rowlist))])]

length(grep("Aut", a$enrichment[-which(a$X %in% a$X[unique(unlist(rowlist))])]))

length(grep("Control", a$enrichment[-which(a$X %in% a$X[unique(unlist(rowlist))])]))


```
