## Load libraries
```{r cache=FALSE, include=TRUE, echo=FALSE, message=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE, cache=FALSE, fig.path=paste0('./Figures_', Sys.Date(), '/')) 
# utility
library(shiny)
library(tibble)
library(data.table)
library(devtools)
library(knitr)
library(tidyr)
library(reshape2)
library(dplyr)
# visualization
library(ggplot2)
library(pander)
library(DT)
library(gridExtra)
library(adegraphics)
library(stats)
#library(smatr)
library(caret)
library(randomForest)
library(ROCR)

library(exactRankTests)
library(nlme)
#library(compositions)
#git clone from: https://github.com/FrederickHuangLin/ANCOM.git
#source("C:/Users/ctata/ANCOM/scripts/ancom_v2.1.R")
#library(ggpubr)

# data analysis

#Need biocondutor (installation method may have changed recently)
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("vegan")
library(vegan)
#BiocManager::install("metagenomeSeq")
library(metagenomeSeq)
#BiocManager::install("DESeq2")
library(DESeq2)
#BiocManager::install("biomformat")
library(biomformat)
#BiocManager::install("phyloseq")
library(phyloseq)

sgColorPalette = c("#84CF04","#01B5BB","#E50E63","#6D7272","#8F389E",
                     "#DF8236","#036B6B","#F1BA2F","#9F832D","#94E804",
                     "#01D4DB","#FAC131","#B0B8B8","#F08C3A","#FF106E",
                     "#B948CC","#05B5B5","#CFAA3A")
```

## All parameters
### File paths
```{r filepaths}
# Assumes your working directory is in M3_Phyloseq_Analysis/scripts/

# Loading phyloseq object. In filtering_normalization.Rmd we cleaned the mapping file, removed samples that were breastfed or whose phenotype
# could not be confirmed, and changed survey answers into numerals where appropriate
ps <- readRDS("../data/Normalized/ps_DeSeq_pass_min_postDD_min0.03.rds")

## Output data location (subject to change)
#output_data=paste0(M3folder_loc, 'Data/V4/180808_ASVdata4/OutputData_Agefiltered/')
output_data <- "../results/"

```

### Cutoff parameters
```{r cutoff}
# chisquare test cutoff (for diet questionnare results significance)
chisq_cut=0.05
```


# Summarize metadata

```{r sum_meta_fun}
# chisquared test function
run_chisq_test <- function(ps, metacol){
  # ps: phyloseq object
  # metacol: metadata column name to test
  metaDF <- data.frame(sample_data(ps))
  # remove NAs, for some reason, some NA are recorded as a text!
  submetaDF=metaDF[!is.na(metaDF[, metacol]), ]
  submetaDF=submetaDF[submetaDF[, metacol]!='NA', ]
  submetaDF=submetaDF[submetaDF[, metacol]!='', ]   # also remove blank
  # chisquared test
  chisq_res=chisq.test(table(submetaDF[, metacol], submetaDF[, 'phenotype']))
  # extract results
  resDT=data.table(chisq_res$observed)
  # dcast for printing
  resDT <- data.table(dcast(resDT, V1 ~ V2, value.var='N'))
  resDT <- resDT[, testvar:=metacol]
  resDT <- resDT[, chisq_p:=chisq_res$p.value]
  
  return(resDT[, list(testvar, category=V1, A, N, chisq_p)])
}

# composition plot function
plot_composition <- function(chisq_resDT, var_name){
  # chisq_resDT: 4 columns. testvar, category, A, N, chisq_p
  plotDT=melt(chisq_resDT, id.vars=c('testvar', 'category', 'chisq_p'))
  p=ggplot(data=plotDT[plotDT$testvar==var_name], aes(x=variable, y=value, fill=category))+
    geom_bar(stat="identity")+
    xlab('')+ylab('Number of sample')+
    ggtitle(var_name)+
    theme_minimal()+
    theme(legend.title=element_blank(), legend.position="bottom", axis.text.x=element_text(vjust=1, hjust=1))+
    scale_fill_manual(values=sgColorPalette)
  print(p)
}
```

```{r}
map <- data.frame(ps@sam_data)
map_levels<-sapply(map, levels)
map_levelscount<-sapply(map_levels, length)
mapnotfac <- names(map_levelscount[which(map_levelscount >= 18)])
for (i in mapnotfac){
  map[,i]<-as.character(map[,i])
}
metadata_ok <- map
```

### Run a Chi squared test on all categorical variables between ASD and NT samples
```{r sex, fig.width=5, fig.height=4}

#now let's only run the categorical values for chi square and remove the first column which are not metadata 
fac_cat<-names(Filter(is.factor, metadata_ok))
#removing the first 13 columns, since it's not metadata and the last one which is phenotype  
fac_cat<-fac_cat[-c(1:13, length(fac_cat))]
#finally removing the ones that were only asked for the children with ASD, or only have one factor & NA, or only present in one phen
fac_cat<-fac_cat[-which(fac_cat %in% c("Behavior.video.submitted..M3.","Language.ability.and.use","Conversation.ability","Understands.speech","Plays.imaginatively.when.alone","Plays.imaginatively.with.others","Plays.in.a.group.with.others","Eye.contact.finding","Childhood.behavioral.development.finding","Repetitive.motion","Picks.up.objects.to.show.to.others","Sleep.pattern.finding","Response.to.typical.sounds","Self.injurious.behavior.finding","Gastrointestinal.problems..M3.", "Imitation.behavior", "Other.stool.sample.collection.method.explained..M3.", "Flu.shot.in.the.last..MFlu.shot.in.the.last..M3.", "Pica.disease", "Additional.info.affecting.microbiome..M3.", "Dietary.restrictions.details..M3.", "Pet.bird"))]



#Also remove the ones with only one factor (no chi-square possible)
#now running the chisquare on all categorical values 
chisquare_p.val=c()
names_chisquare_p.val=c()
all_chisquare=list()
chi_list<-names(map_levelscount)[map_levelscount > 1]
chi_list <-chi_list[-c(1:9)]
chi_list<-chi_list[-which(chi_list %in% c("Behavior.video.submitted..M3.","Language.ability.and.use","Conversation.ability","Understands.speech","Plays.imaginatively.when.alone","Plays.imaginatively.with.others","Plays.in.a.group.with.others","Eye.contact.finding","Childhood.behavioral.development.finding","Repetitive.motion","Picks.up.objects.to.show.to.others","Sleep.pattern.finding","Response.to.typical.sounds","Self.injurious.behavior.finding","Gastrointestinal.problems..M3.", "Imitation.behavior", "Other.stool.sample.collection.method.explained..M3.", "Flu.shot.in.the.last..MFlu.shot.in.the.last..M3.", "Pica.disease", "Additional.info.affecting.microbiome..M3.", "Dietary.restrictions.details..M3.", "Pet.bird", "Host.disease.status", "phenotype", "LR6.predictions..M3.", "LR10.prediction..M3.", "LR5.prediction..M3."))]

#Run chi on all

for (i in 1:length(chi_list)){
    tmp<-run_chisq_test(ps, chi_list[i])
    chisquare_p.val<-c(chisquare_p.val,min(tmp$chisq_p))
    names_chisquare_p.val<-c(names_chisquare_p.val,chi_list[i])
    all_chisquare[[i]]<-tmp
}
names(chisquare_p.val)<-names_chisquare_p.val
names(all_chisquare) <-chi_list

# p-value correction
chisquare_p.val<-p.adjust(chisquare_p.val)
sig_catvar <-chisquare_p.val[chisquare_p.val < 0.05]

length(sig_catvar) #18
sig_catvar

#vizualisation of the results 
#select only the signififcant ones
all_chisquare<-all_chisquare[names(all_chisquare) %in% names(sig_catvar)]
#save this into a csv
write.csv(format(sig_catvar, digits=2), file=paste0(output_data,"Xsqr_05.csv"), quote=F)
# plot one example out of 29 
plot_composition(all_chisquare[1], names(all_chisquare)[1])

#perhaps not surprisingly , many of variables that are significantly different between ASD and NT conditions are lifestyle choices presumably made by the parents. Things like dietary regime, dietary restrictions, and dietary supplements
```




#Run a wilcoxon rank sum test or t test as appropriate on all the numerical variables between ASD and NT cases
```{r}
map_num<-sapply(map, is.numeric)
num_cat <- colnames(map[,as.vector(which(map_num == TRUE))])
num_cat <- num_cat[-c(1:4)]
num_cat <- num_cat[-which(num_cat %in% c("Language.ability.and.use", "Conversation.ability", "Understands.speech", "Mobile.Autism.Risk.Assessment.Score", "Number.of.small.pet.herbivores", "Number.of.small.pet.rodents", "phenotype_num", "Number.of.pet.birds"))]


#Testing to see if numerical variables differ in means using wilcox or t-tests
tmp<-shapiro.test(unlist(map[,num_cat[1]]))
  if (tmp$p.value <= 0.05) {
    res<-wilcox.test(as.formula(paste(num_cat[1], "~ phenotype")), data=map, var.equal = FALSE)
    tab<-tibble(num_cat[1], res$p.value, "wilcox")
    colnames(tab) <- c("Var", "p.value", "Type")
    numerical_test_btwn_pheno <- tab
  }else{
    res<-t.test(as.formula(paste(num_cat[1], "~ phenotype")), data=map)
    tab<-tibble(num_cat[1], res$p.value, "t.test")
    colnames(tab) <- c("Var", "p.value", "Type")
    numerical_test_btwn_pheno <- tab
  }


for (i in num_cat[c(-1)]) {
  tmp<-shapiro.test(unlist(map[,i]))
  if (tmp$p.value <= 0.05) {
    res<-wilcox.test(as.formula(paste(i, "~ phenotype")), data=map, var.equal = FALSE)
    tab<-tibble(i, res$p.value, "wilcox")
    colnames(tab) <- c("Var", "p.value", "Type")
    numerical_test_btwn_pheno <-rbind(numerical_test_btwn_pheno, tab)
  }else{
    res<-t.test(as.formula(paste(i, "~ phenotype")), data=map)
    tab<-tibble(i, res$p.value, "t.test")
    colnames(tab) <- c("Var", "p.value", "Type")
    numerical_test_btwn_pheno <-rbind(numerical_test_btwn_pheno, tab)
  }
}


numerical_test_btwn_pheno$p.value<-p.adjust(numerical_test_btwn_pheno$p.value)
sig_numvar<-numerical_test_btwn_pheno[which(numerical_test_btwn_pheno$p.value <= 0.05),]

num_confounds_all <- sig_numvar$Var



```



### Racial.group
```{r race, fig.width=5, fig.height=4}
# print number table
table(sample_data(ps)$Racial.group, sample_data(ps)$Biological.sex)
# run chisquared test
race=run_chisq_test(ps, 'Racial.group')
# print results
pander(race)
# plot
#plot_composition(race, 'Racial.group')
# % table
race_prop=prop.table(as.matrix(race[, .(A, N)]), margin=2)*100
row.names(race_prop) <- race$category
pander(race_prop)

# write
write.csv(race_prop, file=paste0(output_data, 'Race.csv'))
```



### Check whether there is a large age discrepancy between ASD and NT samples. 
```{r age, fig.width=5, fig.height=4}
# make sure it is numeric
sample_data(ps)$Age..months. <- as.numeric(sample_data(ps)$Age..months.)

# plot
ggplot(data=sample_data(ps), aes(x=phenotype, y=Age..months., fill=phenotype))+
  geom_boxplot(width=0.7, outlier.colour='white')+
  geom_jitter(size=1, position=position_jitter(width=0.1))+
  xlab('')+ylab('Age (months)')+
  scale_fill_manual(values=sgColorPalette)+
  theme_minimal()

# run tests to check significance
shapiro.test(sample_data(ps)$Age..months.) #not normal we need a reanking test
wilcox.test(Age..months. ~ phenotype, data=data.frame(sample_data(ps)), var.equal=FALSE)


#We see the NT have a much higher age range - this is likely to do with the way that we phrased the eligibility criteria - ASD child between the ages of 2 and 10 and a NT within 2 years of age
```