---
title: "plotting_taxa_abundance"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(reshape2)
library(phyloseq)
library(DT)
library(data.table)

sgColorPalette = c("#01B5BB","#84CF04","#E50E63","#6D7272","#8F389E",
                     "#DF8236","#036B6B","#F1BA2F","#9F832D","#94E804",
                     "#01D4DB","#FAC131","#B0B8B8","#F08C3A","#FF106E",
                     "#B948CC","#05B5B5","#CFAA3A")


#C9EAFF
```

# Visualizing results from DESeq, MetagenomeSeq, and Ancom
### Read in results from above tests
```{r}
output_data <- "../results/"
sig_taxa <- read.table(paste0(output_data, "significant_taxa_timeseries.tsv"), sep = "\t", header = T)

#Load for plotting purposes:
ps_tss <- readRDS(paste0(output_data, "Normalized/ps_tss_pass_min_postDD_min0.03.rds"))
ps_des <- readRDS(paste0(output_data, "Filtered/ps_deseq_friedfilt.rds"))

# fix taxonomic assignments
strain_assignment <- read.csv("../data/INRD17_0415_combinedRuns_multi_hits.csv")
strain_assignment <- strain_assignment[!duplicated(strain_assignment$ss_id), ]
rownames(strain_assignment) <- strain_assignment$ss_id
strain_names <- taxa_names(ps_tss)[grepl("t__", taxa_names(ps_tss))]
names_split <- strsplit(as.character(strain_assignment[strain_names, ]$display_name), split = ".", fixed = T)
genus <- unlist(lapply(names_split, function(x) return(x[[1]])))
species <- unlist(lapply(names_split, function(x) return(x[[2]])))
ps_tss@tax_table[strain_names, 6] <- paste("g__", genus, sep = "")
ps_tss@tax_table[strain_names, 7] <- paste("s__", species, sep = "")

species_names <- as.character(ps_tss@tax_table[, 7])
species_names <- unlist(lapply(strsplit(species_names, "__"), function(x) return(x[length(x)])))
species_names <- paste0("s__", species_names)
ps_tss@tax_table[, 7] <- species_names

#For plots
ps_tss@sam_data$phenotype <- gsub("A", "ASD", ps_tss@sam_data$phenotype)
ps_tss@sam_data$phenotype <- gsub("N", "NT", ps_tss@sam_data$phenotype)

ps_tss@sam_data$phenotype_TD <- gsub("NT", "TD", ps_tss@sam_data$phenotype)

```


```{r visualization_fun}
### functions to plot
make_vis_plots <- function(ps_norm, grouping, tax, plot_type=c('box', 'bar')){
  # ps should be a normalized (DESeq or CSS) phyloseq object
  # grouping should match the column name in the sample_data
  # tax is a taxonomical bin id (ASV) in the counts table to plot
  
  # subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_norm) %in% tax, ps_norm)

  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  #taxa names for plot labels

  species_names <- as.character(ps_tss@tax_table[plot_table$rn, 7])
  genus_names <- as.character(ps_tss@tax_table[plot_table$rn, 6])
  family_names <- as.character(ps_tss@tax_table[plot_table$rn, 5])
  taxa_names <- paste( family_names, genus_names,species_names, sep = "\n")
  plot_table$rn <- taxa_names
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')
  
  # boxplot
  if(plot_type=='box'){
    ggplot(data=plot_table, aes(x=Within.study.sampling.date, y = value, fill=Group)) + 
      geom_boxplot(outlier.color=NA) +
      geom_jitter(position=position_jitterdodge(0.2), cex=1.5, color="gray44") + 
      labs(title ="Figure 3: Relative Abundance Counts of Taxa Significant in Two Independent Methods", x='', y ='Proportional counts, log scale') + 
      scale_y_log10() + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  } else if (plot_type=='bar'){
    plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title ="Figure 3: Relative Abundance Counts of Taxa Significant in Two Independent Methods", x='', y ='Proportional counts, 0 to 1 scale') +
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal(base_size = 13) + facet_wrap(~rn, scales='free', ncol=4)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  }
}

make_vis_plots_des <- function(ps_norm, grouping, tax, plot_type=c('box', 'bar')){
  # ps should be a normalized (DESeq or CSS) phyloseq object
  # grouping should match the column name in the sample_data
  # tax is a taxonomical bin id (ASV) in the counts table to plot
  
  # subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_norm) %in% tax, ps_norm)

  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  #plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  #taxa names for plot labels

  species_names <- as.character(ps_tss@tax_table[plot_table$rn, 7])
  genus_names <- as.character(ps_tss@tax_table[plot_table$rn, 6])
  family_names <- as.character(ps_tss@tax_table[plot_table$rn, 5])
  taxa_names <- paste( family_names, genus_names,species_names, sep = "\n")
  plot_table$rn <- taxa_names
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')

  # boxplot
  if(plot_type=='box'){
    ggplot(data=plot_table, aes(x=Within.study.sampling.date, y = value, fill=Group)) + 
      geom_boxplot(outlier.color=NA) +
      geom_jitter(position=position_jitterdodge(0.2), cex=1.5, color="gray44") + 
      labs(title ="Figure 3: Relative Abundance Counts of Taxa Significant in Two Independent Methods", x='', y ='Proportional counts, log scale') + 
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  } else if (plot_type=='bar'){
    plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title ="Figure 3: Relative Abundance Counts of Taxa Significant in Two Independent Methods", x='', y ='DEseq2 normalized counts') +
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal() + facet_wrap(~rn, scales='free', ncol=3)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
  }
}

```

## Both DESeq and metagenomeseq timeseries
```{r fig.width=12, fig.height=10}
keep <- grepl("DESEQ", sig_taxa$method) & grepl("ZIG", sig_taxa$method)
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$asv, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$asv, 'bar'))
} else {
  print('no combined timeseries significant taxa')
}

#for some reason, only first 11 are printing from the code above despite there being 13, will do so manually
#print(make_vis_plots(ps_tss, 'phenotype', sig_taxa[keep, ]$asv[10], 'bar'))
keep1 <-keep
```

## Both ZIG and ANCOM timeseries
```{r fig.width=10, fig.height=8.5}
keep <- grepl("ZIG", sig_taxa$method) & grepl("ANCOM", sig_taxa$method)
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$asv, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$asv, 'bar'))
} else {
  print('no combined timeseries significant taxa')
}

#Best ones
print(make_vis_plots(ps_tss, 'phenotype_TD', c(sig_taxa[keep, ]$asv, sig_taxa[keep1, ]$asv[c(1,6,8)]), 'bar'))
#For paper
print(make_vis_plots(ps_tss, 'phenotype_TD', c(sig_taxa[keep,]$asv, sig_taxa[keep1,]$asv), 'bar'))
#print(make_vis_plots(ps_des, 'phenotype', c(sig_taxa[keep,]$asv, sig_taxa[keep1,]$asv), 'bar'))
tax <-  c(sig_taxa[keep,]$asv, sig_taxa[keep1,]$asv)
grouping <- 'phenotype_TD'
# subset phyloseq object to select ASV of interest
  ps_filt=prune_taxa(taxa_names(ps_tss) %in% tax, ps_tss)

  # get normalized counts
  plot_table<-data.table(otu_table(ps_filt), keep.rownames=TRUE)[rn %in% tax]
  # add very small value, min/100000 to 0
  plot_table <- melt(plot_table, id.vars='rn')
  plot_table$value <- plot_table$value+min(plot_table[value!=0]$value)/100000
  
  # add metadata
  groupDT=data.table(data.frame(sample_data(ps_filt)[, c(grouping, 'Within.study.sampling.date')]), keep.rownames=TRUE)
  setnames(groupDT, 'rn', 'variable')
  plot_table <- merge(plot_table, groupDT, by='variable', all.x=TRUE)
  
  #taxa names for plot labels

  species_names <- as.character(ps_tss@tax_table[plot_table$rn, 7])
  genus_names <- as.character(ps_tss@tax_table[plot_table$rn, 6])
  family_names <- as.character(ps_tss@tax_table[plot_table$rn, 5])
  taxa_names <- paste( family_names, genus_names,species_names, sep = "\n")
  plot_table$rn <- taxa_names
  
  # change variable to general name
  setnames(plot_table, grouping, 'Group')
  
  plot_table$rn<-factor(plot_table$rn, levels = c("f__Bacteroidaceae\ng__Bacteroides\ns__thetaiotaomicron",   "f__Lachnospiraceae\ng__Blautia\ns__unclassified",  "f__Lachnospiraceae\ng__Blautia_A\ns__unclassified" ,  "f__Lachnospiraceae\ng__unclassified\ns__unclassified", "f__Oscillospiraceae\ng__unclassified\ns__unclassified", "f__Anaerovoracaceae\ng__unclassified\ns__unclassified", "f__Erysipelotrichaceae\ng__Holdemania\ns__unclassified",  "f__Christensenellaceae\ng__borkfalki\ns__ceftriaxensis", "f__Streptococcaceae\ng__Streptococcus\ns__unclassified", "f__Lachnospiraceae\ng__Blautia_A\ns__wexlerae" ,"f__Pasteurellaceae\ng__Haemophilus_D\ns__unclassified"  ))
  plot_table2 <- plot_table[, list(mean_ct=mean(value), sem=sd(value)/sqrt(.N)), by=c('Group', 'Within.study.sampling.date', 'rn')]
    
    ggplot(data=plot_table2, aes(x=Within.study.sampling.date, y =mean_ct, fill=Group)) + 
      geom_bar(stat='identity', position=position_dodge()) +
      geom_errorbar(aes(ymin=mean_ct-sem, ymax=mean_ct+sem), width=0.2, position=position_dodge(0.9))+ 
      labs(title ="Figure 3: Relative Abundance Counts of Taxa Significant in Two Independent Methods", x='', y ='Proportional counts, 0 to 1 scale') +
      scale_fill_manual(values=sgColorPalette)+
      theme_minimal(base_size = 13) + facet_wrap(~rn, scales='free', ncol=4)+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))

```
## Other variables directionality
```{r}
#Best ones
print(make_vis_plots(ps_tss, 'Dietary.supplement', sig_taxa[keep, ]$asv[2], 'bar'))
print(make_vis_plots(ps_tss, 'GI.symptoms.within.3.months..M3.', sig_taxa[keep, ]$asv[2], 'bar'))
print(make_vis_plots(ps_tss, 'Lactose.intolerance', sig_taxa[keep, ]$asv[3], 'bar'))
print(make_vis_plots(ps_tss, 'Multivitamin', sig_taxa[keep, ]$asv[3], 'bar'))
print(make_vis_plots(ps_tss, 'Dietary.restrictions..M3.', sig_taxa[keep, ]$asv[3], 'bar'))
```

## Both DESeq and ANCOM timeseries
```{r fig.width=12, fig.height=10}
keep <- grepl("DESeq2", sig_taxa$method) & grepl("ANCOM", sig_taxa$method)
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$asv, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$asv, 'bar'))
} else {
  print('no combined timeseries significant taxa')
}
```


### At least 1 timepoint in both DESeq and metagenomeSeq
```{r fig.width=12, fig.height=7, eval = F, echo = F}
keep_deseq <- !is.na(sig_taxa$deseq_P1_adjp) | !is.na(sig_taxa$deseq_P2_adjp) | !is.na(sig_taxa$deseq_P3_adjp)
keep_metagenome <- !is.na(sig_taxa$mtgseq_P1_adjp) | !is.na(sig_taxa$mtgseq_P2_adjp) | !is.na(sig_taxa$mtgseq_P3_adjp)
keep <- keep_deseq & keep_metagenome
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype_TD', sig_taxa[keep, ]$rn, 'bar'))
} else {
  print('no DESeq timeseries significant taxa')
}
```


### DESeq all 3 timepoints
```{r visualization_sig_deseq, fig.width=12, fig.height=4, eval = F, echo = F}
## plot
# common by deseq
com_deseq_taxa = sig_taxa[!is.na(sig_taxa$deseq_P1_adjp) & !is.na(sig_taxa$deseq_P2_adjp) & !is.na(sig_taxa$deseq_P3_adjp), ]

if(nrow(com_deseq_taxa)>0){
  print(make_vis_plots(ps_tss, 'phenotype', com_deseq_taxa$rn, 'box'))
  print(make_vis_plots(ps_tss, 'phenotype', com_deseq_taxa$rn, 'bar'))
} else {
  print('no common DESeq significant taxa')
}
```

### metagenomeSeq all 3 timepoints
```{r visualization_sig_mtgseq, fig.width=12, fig.height=10, eval = F, echo = F}
# common by metagenomeseq
com_mtgseq_taxa=sig_taxa[!is.na(sig_taxa$mtgseq_P1_adjp) & !is.na(sig_taxa$mtgseq_P2_adjp) & !is.na(sig_taxa$mtgseq_P3_adjp), ]

if(nrow(com_mtgseq_taxa)>0){
  print(make_vis_plots(ps_tss, 'phenotype', com_mtgseq_taxa$rn, 'box'))
} else {
  print('no common metagenomeSeq significant taxa')
}
```


### DESeq at least 2 timepoints
```{r visualization_sig_deseq_ts, fig.width=12, fig.height=10, eval = F, echo = F}
# deseq timeseries
keep <- data.frame(!is.na(sig_taxa$deseq_P1_adjp), !is.na(sig_taxa$deseq_P2_adjp), !is.na(sig_taxa$deseq_P3_adjp))
keep <- apply(keep, 1, function(x) sum(x)>1)
if(sum(keep)>0){
  print(make_vis_plots(ps_tss, 'phenotype', sig_taxa[keep, ]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', sig_taxa[keep, ]$rn, 'bar'))
} else {
  print('no DESeq timeseries significant taxa')
}
```



### Meta_genome at least 2 timepoints
```{r visualization_sig_mtg_ts, fig.width=12, fig.height=5, eval = F, echo = F}
keep <- data.frame(!is.na(sig_taxa$mtgseq_P1_adjp), !is.na(sig_taxa$mtgseq_P2_adjp), !is.na(sig_taxa$mtgseq_P3_adjp))
keep <- apply(keep, 1, function(x) sum(x)>1)
# mtgseq timeseries
if(nrow(sig_taxa[!is.na(sig_taxa$mtgseq_timeseries_adjp),])>0){
  print(make_vis_plots(ps_tss, 'phenotype', sig_taxa[keep,]$rn, 'box'))
  # plot bar as well
  print(make_vis_plots(ps_tss, 'phenotype', sig_taxa[keep,]$rn, 'bar'))
} else {
  print('no Mtgseq timeseries significant taxa')
}

```


### t__186843 more visible after antibiotics
```{r}
abund <- ps_tss@otu_table["t__186843", ]
ps_tss@sam_data$Minimum.time.since.antibiotics
plot(ps_tss@sam_data$Minimum.time.since.antibiotics, abund)
```

